// @sap/cds-compiler/lib/gen/CdlParser.js
// Parser generated by redepage v0.1.12
'use strict;'
const { XsnSource, XsnArtifact, XsnName } = require( '../compiler/xsn-model' );
const AstBuildingParser = require('../parsers/AstBuildingParser');
const keywords = {
  abstract: true,
  action: true,
  actions: true,
  all: false,
  and: true,
  annotate: true,
  annotation: true,
  any: false,
  array: true,
  as: false,
  asc: true,
  aspect: true,
  association: true,
  between: true,
  by: false,
  case: false,
  cast: false,
  columns: true,
  composition: true,
  context: true,
  cross: true,
  current: true,
  default: true,
  define: true,
  definitions: true,
  desc: true,
  distinct: true,
  element: true,
  elements: true,
  else: true,
  end: true,
  entity: true,
  enum: true,
  escape: true,
  event: true,
  exact: true,
  except: true,
  excluding: true,
  exists: false,
  extend: true,
  false: false,
  first: true,
  floating: true,
  following: true,
  from: false,
  full: true,
  function: true,
  group: true,
  having: true,
  in: false,
  inner: true,
  intersect: true,
  into: true,
  is: true,
  join: true,
  key: false,
  last: true,
  left: true,
  like: true,
  limit: true,
  localized: true,
  many: true,
  masked: true,
  minus: true,
  mixin: true,
  namespace: true,
  new: true,
  not: false,
  null: false,
  nulls: true,
  of: false,
  offset: true,
  on: false,
  one: true,
  or: true,
  order: true,
  outer: true,
  over: true,
  parameters: true,
  partition: true,
  preceding: true,
  projection: true,
  redirected: true,
  returns: true,
  right: true,
  row: true,
  rows: true,
  select: false,
  service: true,
  some: false,
  stored: true,
  then: true,
  to: true,
  true: false,
  type: true,
  unbounded: true,
  union: true,
  up: true,
  using: true,
  variable: true,
  view: true,
  virtual: true,
  when: true,
  where: false,
  with: false,
};
const table = [
  'start',
  /* 1 */ {
    namespace: [2,16],
    using: [2,19],
    '@': [2,10], type: '@', view: '@', event: '@', action: '@', aspect: '@', define: '@', entity: '@', extend: '@', context: '@', service: '@', abstract: '@', annotate: '@', function: '@', annotation: '@',
    '': 3
  },
  /* 2 */ {
    ';': ['c',1],
    EOF: 3,
    Id: ['g',1], '@': 'Id',
  },
  /* 3 */ ['m',0,'EOF'],
  'artifactsBlock',
  /* 5 */ ['m',6,'{'],
  /* 6 */ { Id: [7,10], '@': 'Id', '': 8 },
  /* 7 */ {
    ';': ['c',6],
    '}': 8,
    Id: ['g',6], '@': 'Id',
  },
  /* 8 */ ['m',0,'}'],
  'artifactDefOrExtend',
  /* 10 */ { '@': [10,777], '': 11 },
  /* 11 */ {
    define: ['ck',12],
    type: 12, view: 12, event: 12, action: 12, aspect: 12, entity: 12, context: 12, service: 12, abstract: 12, function: 12, annotation: 12,
    extend: ['ck',13],
    annotate: ['ck',14],
  },
  /* 12 */ {
    service: [0,44],
    context: [0,49],
    annotation: [0,54],
    type: [0,59],
    aspect: [0,64], abstract: 'aspect',
    entity: [0,77],
    view: [0,91],
    event: [0,103],
    action: [0,115],
    function: [0,121],
  },
  /* 13 */ {
    Id: [0,231],
    service: [0,253,1],
    context: [0,260,1],
    type: [0,267,1],
    aspect: [0,279,1], entity: 'aspect',
    projection: [0,290,1],
  },
  /* 14 */ [0,218],
  'namespaceDeclaration',
  /* 16 */ ['mk',17,'namespace'],
  /* 17 */ [0,36],
  'usingDeclaration',
  /* 19 */ ['mk',20,'using'],
  /* 20 */ {
    from: ['ck',21],
    Id: [22,32],
    '{': 24,
  },
  /* 21 */ ['m',0,'String'],
  /* 22 */ { from: ['ck',23], '': 0 },
  /* 23 */ ['m',0,'String'],
  /* 24 */ ['m',25,'{'],
  /* 25 */ { Id: 26 },
  /* 26 */ { Id: [27,32], '': 28 },
  /* 27 */ {
    ',': ['c',26],
    '}': 28,
  },
  /* 28 */ ['m',29,'}'],
  /* 29 */ { from: ['ck',30], '': 0 },
  /* 30 */ ['m',0,'String'],
  'usingProxy',
  /* 32 */ [33,40],
  /* 33 */ { as: ['ck',34], '': 0 },
  /* 34 */ ['mi',0],
  'namePath',
  /* 36 */ ['mi',37],
  /* 37 */ { '.': ['c',38], '': 0 },
  /* 38 */ ['miA',37],
  'simplePath',
  /* 40 */ ['mi',41],
  /* 41 */ { '.': ['c',42], '': 0 },
  /* 42 */ ['miA',41],
  'serviceDef',
  /* 44 */ ['mk',45,'service'],
  /* 45 */ [46,36],
  /* 46 */ { '@': [46,783], '': 47 },
  /* 47 */ { '{': [0,5], '': 0 },
  'contextDef',
  /* 49 */ ['mk',50,'context'],
  /* 50 */ [51,36],
  /* 51 */ { '@': [51,783], '': 52 },
  /* 52 */ { '{': [0,5], '': 0 },
  'annotationDef',
  /* 54 */ ['mk',55,'annotation'],
  /* 55 */ [56,36],
  /* 56 */ { '@': [56,783], '': 57 },
  /* 57 */ [0,349],
  'typeDef',
  /* 59 */ ['mk',60,'type'],
  /* 60 */ [61,36],
  /* 61 */ { '@': [61,783], '': 62 },
  /* 62 */ [0,349],
  'aspectDef',
  /* 64 */ {
    aspect: ['ck',66],
    abstract: ['ck',65],
  },
  /* 65 */ ['mk',66,'entity'],
  /* 66 */ [67,36],
  /* 67 */ { '@': [67,783], '': 68 },
  /* 68 */ {
    '{': [69,160],
    ':': ['c',70],
    '': 0
  },
  /* 69 */ { actions: [0,127], '': 0 },
  /* 70 */ [0,72],
  'aspectColonSpec',
  /* 72 */ { Id: [73,40], '': 74 },
  /* 73 */ { ',': ['c',72], '': 74 },
  /* 74 */ { '{': [75,160], '': 0 },
  /* 75 */ { actions: [0,127], '': 0 },
  'entityDef',
  /* 77 */ ['mk',78,'entity'],
  /* 78 */ [79,36],
  /* 79 */ { '@': [79,783], '': 80 },
  /* 80 */ { '(': [81,144], '': 81 },
  /* 81 */ {
    ':': ['c',82],
    '{': 85,
    as: ['ck',86],
  },
  /* 82 */ { Id: 83 },
  /* 83 */ { Id: [84,40], '': 85 },
  /* 84 */ {
    ',': ['c',83],
    '{': 85,
  },
  /* 85 */ [89,160],
  /* 86 */ {
    '(': [89,461], select: '(',
    projection: [87,455],
  },
  /* 87 */ { group: [88,610], where: 'group', having: 'group', '': 88 },
  /* 88 */ { limit: [89,622], order: 'limit', '': 89 },
  /* 89 */ { actions: [0,127], '': 0 },
  'viewDef',
  /* 91 */ ['mk',92,'view'],
  /* 92 */ [93,36],
  /* 93 */ { '@': [93,783], '': 94 },
  /* 94 */ {
    '(': [100,144],
    with: ['ck',95],
    '': 100
  },
  /* 95 */ ['mk',96,'parameters'],
  /* 96 */ [97,149],
  /* 97 */ { ',': ['c',98], '': 99 },
  /* 98 */ [97,149],
  /* 99 */ ['g',100],
  /* 100 */ ['mk',101,'as'],
  /* 101 */ [0,461],
  'eventDef',
  /* 103 */ ['mk',104,'event'],
  /* 104 */ [105,36],
  /* 105 */ { '@': [105,783], '': 106 },
  /* 106 */ {
    '{': [0,160],
    ':': ['c',107],
  },
  /* 107 */ {
    '{': [0,160],
    Id: [108,40],
    projection: [0,455,1],
  },
  /* 108 */ { ',': 109, '{': 109, '': 113 },
  /* 109 */ { ',': ['c',110], '': 112 },
  /* 110 */ { Id: [111,40], '': 112 },
  /* 111 */ {
    ',': ['c',110],
    '{': 112,
  },
  /* 112 */ [0,160],
  /* 113 */ { '@': [113,777], '': 0 },
  'actionMainDef',
  /* 115 */ ['mk',116,'action'],
  /* 116 */ [117,36],
  /* 117 */ { '@': [117,783], '': 118 },
  /* 118 */ [119,144],
  /* 119 */ { returns: [0,156], '': 0 },
  'functionMainDef',
  /* 121 */ ['mk',122,'function'],
  /* 122 */ [123,36],
  /* 123 */ { '@': [123,783], '': 124 },
  /* 124 */ [125,144],
  /* 125 */ [0,156],
  'actionsBlock',
  /* 127 */ ['mk',128,'actions'],
  /* 128 */ ['m',129,'{'],
  /* 129 */ { Id: [130,133], '@': 'Id', '': 131 },
  /* 130 */ {
    ';': ['c',129],
    '}': 131,
    Id: ['g',129], '@': 'Id',
  },
  /* 131 */ ['m',0,'}'],
  'boundActionFunctionDef',
  /* 133 */ { '@': [133,777], '': 134 },
  /* 134 */ {
    action: ['ck',135],
    function: ['ck',139],
  },
  /* 135 */ ['mi',136],
  /* 136 */ { '@': [136,783], '': 137 },
  /* 137 */ [138,144],
  /* 138 */ { returns: [0,156], '': 0 },
  /* 139 */ ['mi',140],
  /* 140 */ { '@': [140,783], '': 141 },
  /* 141 */ [142,144],
  /* 142 */ [0,156],
  'paramsList',
  /* 144 */ ['m',145,'('],
  /* 145 */ { Id: [146,149], '@': 'Id', '': 147 },
  /* 146 */ {
    ',': ['c',145],
    ')': 147,
  },
  /* 147 */ ['m',0,')'],
  'paramDef',
  /* 149 */ { '@': [149,777], '': 150 },
  /* 150 */ ['miA',151],
  /* 151 */ { '@': [151,783], '': 152 },
  /* 152 */ {
    '{': [153,160],
    ':': ['c',154],
  },
  /* 153 */ { not: [0,448], null: 'not', '': 0 },
  /* 154 */ [0,360],
  'returnsSpec',
  /* 156 */ ['mk',157,'returns'],
  /* 157 */ { '@': [157,777], '': 158 },
  /* 158 */ [0,360],
  'elementsBlock',
  /* 160 */ ['m',161,'{'],
  /* 161 */ { Id: [162,165], '@': 'Id', '': 163 },
  /* 162 */ {
    ';': ['c',161],
    '}': 163,
    Id: ['g',161], '@': 'Id',
  },
  /* 163 */ ['m',0,'}'],
  'elementDef',
  /* 165 */ { '@': [165,777], '': 166 },
  /* 166 */ { virtual: ['ck',167,1], '': 167 },
  /* 167 */ { key: ['ck',168], '': 168 },
  /* 168 */ { masked: ['ck',169,1], '': 169 },
  /* 169 */ { element: ['ck',170,1], '': 170 },
  /* 170 */ ['mi',171],
  /* 171 */ { '@': [171,783], '': 172 },
  /* 172 */ {
    '{': [173,160],
    ':': ['c',174],
    '': 175
  },
  /* 173 */ { not: [175,448], null: 'not', '': 175 },
  /* 174 */ [175,360],
  /* 175 */ { '=': ['c',176], '': 0 },
  /* 176 */ [177,650],
  /* 177 */ { stored: ['ck',178], '': 178 },
  /* 178 */ ['g',179],
  /* 179 */ { '@': [179,777], '': 0 },
  'enumSymbolsBlock',
  /* 181 */ ['mk',182,'enum'],
  /* 182 */ ['m',183,'{'],
  /* 183 */ { Id: [184,187], '@': 'Id', '': 185 },
  /* 184 */ {
    ';': ['c',183],
    '}': 185,
  },
  /* 185 */ ['m',0,'}'],
  'enumSymbolDef',
  /* 187 */ { '@': [187,777], '': 188 },
  /* 188 */ ['mi',189],
  /* 189 */ { '@': [189,777], '': 190 },
  /* 190 */ { '=': ['c',191], '': 0 },
  /* 191 */ {
    String: ['c',193],
    Number: ['c',193],
    '+': ['c',192], '-': '+',
    Id: [193,774], '#': 'Id', QuotedLiteral: 'Id',
  },
  /* 192 */ ['m',193,'Number'],
  /* 193 */ ['g',194],
  /* 194 */ { '@': [194,777], '': 0 },
  'foreignKeysBlock',
  /* 196 */ ['m',197,'{'],
  /* 197 */ { Id: [198,201], '@': 'Id', '': 199 },
  /* 198 */ {
    ',': ['c',197],
    '}': 199,
  },
  /* 199 */ ['m',0,'}'],
  'foreignKeyDef',
  /* 201 */ { '@': [201,777], '': 202 },
  /* 202 */ [203,40],
  /* 203 */ { as: ['ck',204], '': 0 },
  /* 204 */ ['mi',0],
  'mixinElementDef',
  /* 206 */ ['mi',207],
  /* 207 */ ['m',208,':'],
  /* 208 */ {
    association: ['ck',209],
    composition: ['ck',211],
  },
  /* 209 */ { '[': [210,430], '': 210 },
  /* 210 */ ['mk',213,'to'],
  /* 211 */ { '[': [212,430], '': 212 },
  /* 212 */ ['mk',213,'of'],
  /* 213 */ { one: ['ck',214,1], many: 'one', '': 214 },
  /* 214 */ [215,40],
  /* 215 */ ['mk',216,'on'],
  /* 216 */ [0,642],
  'annotateArtifact',
  /* 218 */ [219,36],
  /* 219 */ {
    ':': ['c',220],
    with: ['ck',225],
    '': 225
  },
  /* 220 */ [221,36],
  /* 221 */ { with: ['ck',222], '': 222 },
  /* 222 */ ['g',223],
  /* 223 */ { '@': [223,777], '': 224 },
  /* 224 */ { '{': [0,323], '': 0 },
  /* 225 */ ['g',226],
  /* 226 */ { '@': [226,777], '': 227 },
  /* 227 */ { '(': [228,310], '': 228 },
  /* 228 */ {
    returns: [0,319],
    '{': [229,323],
    '': 229
  },
  /* 229 */ { actions: [0,298], '': 0 },
  'extendArtifact',
  /* 231 */ [232,36],
  /* 232 */ {
    ':': ['c',233],
    with: ['ck',242],
    '': 239
  },
  /* 233 */ [234,36],
  /* 234 */ { with: ['ck',235], '': 235 },
  /* 235 */ ['g',236],
  /* 236 */ { '@': [236,777], '': 237 },
  /* 237 */ {
    elements: ['ck',238],
    '{': 238,
    enum: [0,181],
    '(': [0,421],
    '': 0
  },
  /* 238 */ [0,333],
  /* 239 */ { '@': [239,777], '': 240 },
  /* 240 */ { '{': [241,333], '': 0 },
  /* 241 */ { actions: [0,127], '': 0 },
  /* 242 */ { '@': [242,777], '': 243 },
  /* 243 */ {
    Id: [244,40],
    elements: ['ck',248,1],
    '{': 248,
    actions: [0,127,1],
    enum: [0,181,1],
    '(': [0,421],
    columns: ['ck',250,1],
    definitions: ['ck',251,1],
    '': 0
  },
  /* 244 */ { ',': ['c',245], '': 246 },
  /* 245 */ [244,40],
  /* 246 */ { '{': [247,333], '': 247 },
  /* 247 */ { actions: [0,127], '': 0 },
  /* 248 */ [249,333],
  /* 249 */ { actions: [0,127], '': 0 },
  /* 250 */ [0,566],
  /* 251 */ [0,5],
  'extendService',
  /* 253 */ ['mk',254,'service'],
  /* 254 */ [255,36],
  /* 255 */ { with: ['ck',256], '': 256 },
  /* 256 */ ['g',257],
  /* 257 */ { '@': [257,777], '': 258 },
  /* 258 */ { '{': [0,5], '': 0 },
  'extendContext',
  /* 260 */ ['mk',261,'context'],
  /* 261 */ [262,36],
  /* 262 */ { with: ['ck',263], '': 263 },
  /* 263 */ ['g',264],
  /* 264 */ { '@': [264,777], '': 265 },
  /* 265 */ { '{': [0,5], '': 0 },
  'extendType',
  /* 267 */ ['mk',268,'type'],
  /* 268 */ [269,36],
  /* 269 */ { with: ['ck',272], '': 270 },
  /* 270 */ { '@': [270,777], '': 271 },
  /* 271 */ { '{': [0,333], '': 0 },
  /* 272 */ { '@': [272,777], '': 273 },
  /* 273 */ {
    Id: [274,40],
    elements: ['ck',277,1],
    '{': 277,
    enum: [0,181,1],
    '(': [0,421],
    '': 0
  },
  /* 274 */ { ',': ['c',275], '': 276 },
  /* 275 */ [274,40],
  /* 276 */ { '{': [0,333], '': 0 },
  /* 277 */ [0,333],
  'extendEntityOrAspect',
  /* 279 */ { aspect: ['ck',280], entity: 'aspect' },
  /* 280 */ [281,36],
  /* 281 */ { with: ['ck',283], '': 282 },
  /* 282 */ { '@': [282,777], '': 287 },
  /* 283 */ { '@': [283,777], '': 284 },
  /* 284 */ {
    Id: [285,40],
    actions: ['g',287,1],
    '': 287
  },
  /* 285 */ { ',': ['c',286], '': 287 },
  /* 286 */ [285,40],
  /* 287 */ { '{': [288,333], '': 288 },
  /* 288 */ { actions: [0,127], '': 0 },
  'extendProjection',
  /* 290 */ ['mk',291,'projection'],
  /* 291 */ [292,36],
  /* 292 */ { with: ['ck',293], '': 293 },
  /* 293 */ ['g',294],
  /* 294 */ { '@': [294,777], '': 295 },
  /* 295 */ { '{': [296,566], '': 296 },
  /* 296 */ { actions: [0,127], '': 0 },
  'annotateActionsBlock',
  /* 298 */ ['mk',299,'actions'],
  /* 299 */ ['m',300,'{'],
  /* 300 */ { Id: [301,304], '@': 'Id', '': 302 },
  /* 301 */ {
    ';': ['c',300],
    '}': 302,
    Id: ['g',300], '@': 'Id',
  },
  /* 302 */ ['m',0,'}'],
  'annotateBoundAction',
  /* 304 */ { '@': [304,777], '': 305 },
  /* 305 */ ['mi',306],
  /* 306 */ { '@': [306,777], '': 307 },
  /* 307 */ { '(': [308,310], '': 308 },
  /* 308 */ { returns: [0,319,1], '': 0 },
  'annotateParamsBlock',
  /* 310 */ ['m',311,'('],
  /* 311 */ { Id: [312,315], '@': 'Id', '': 313 },
  /* 312 */ {
    ',': ['c',311],
    ')': 313,
  },
  /* 313 */ ['m',0,')'],
  'annotateParam',
  /* 315 */ { '@': [315,777], '': 316 },
  /* 316 */ ['mi',317],
  /* 317 */ { '@': [317,777], '': 0 },
  'annotateReturns',
  /* 319 */ ['mk',320,'returns'],
  /* 320 */ { '@': [320,777], '': 321 },
  /* 321 */ { '{': [0,323], '': 0 },
  'annotateElementsBlock',
  /* 323 */ ['m',324,'{'],
  /* 324 */ { Id: [325,328], '@': 'Id', '': 326 },
  /* 325 */ {
    ';': ['c',324],
    '}': 326,
    Id: ['g',324], '@': 'Id',
  },
  /* 326 */ ['m',0,'}'],
  'annotateElement',
  /* 328 */ { '@': [328,777], '': 329 },
  /* 329 */ ['mi',330],
  /* 330 */ { '@': [330,777], '': 331 },
  /* 331 */ { '{': [0,323], '': 0 },
  'extendElementsBlock',
  /* 333 */ ['m',334,'{'],
  /* 334 */ { Id: [335,338], '@': 'Id', '': 336 },
  /* 335 */ {
    ';': ['c',334],
    '}': 336,
    Id: ['g',334], '@': 'Id',
  },
  /* 336 */ ['m',0,'}'],
  'elementDefOrExtend',
  /* 338 */ { '@': [338,777], '': 339 },
  /* 339 */ {
    Id: [0,165], '@': 'Id', key: 'Id',
    extend: ['ck',340,1],
  },
  /* 340 */ { element: ['ck',341,1], '': 341 },
  /* 341 */ ['mi',342],
  /* 342 */ { with: ['ck',345], '': 343 },
  /* 343 */ { '@': [343,777], '': 344 },
  /* 344 */ { '{': [0,333], '': 0 },
  /* 345 */ { '@': [345,777], '': 346 },
  /* 346 */ {
    elements: ['ck',347],
    '{': 347,
    enum: [0,181],
    '(': [0,421],
    '': 0
  },
  /* 347 */ [0,333],
  'typeOrIncludesSpec',
  /* 349 */ {
    '{': [350,160],
    ':': ['c',351],
  },
  /* 350 */ { not: [0,448], null: 'not', '': 0 },
  /* 351 */ {
    '{': [0,360,1], many: '{', type: '{', array: '{', localized: '{', association: '{', composition: '{',
    Id: [352,40],
  },
  /* 352 */ {
    '(': [353,410], ':': '(',
    '@': [0,362], not: '@', enum: '@', null: '@', default: '@',
    ',': 354, '{': 354,
    '': 0
  },
  /* 353 */ { '@': [0,362], not: '@', enum: '@', null: '@', default: '@', '': 0 },
  /* 354 */ { ',': ['c',355], '': 357 },
  /* 355 */ { Id: [356,40], '': 357 },
  /* 356 */ {
    ',': ['c',355],
    '{': 357,
  },
  /* 357 */ [358,160],
  /* 358 */ { not: [0,448], null: 'not', '': 0 },
  'typeExpression',
  /* 360 */ {
    '{': [361,160],
    Id: [362,409],
    type: [362,401,1],
    localized: ['ck',370,1],
    association: ['ck',374,1],
    composition: ['ck',383,1],
    array: ['ck',393,1],
    many: ['ck',394,1],
  },
  /* 361 */ { not: [0,448], null: 'not', '': 0 },
  /* 362 */ { not: [363,448], null: 'not', '': 363 },
  /* 363 */ ['g',364],
  /* 364 */ {
    not: [365,442], null: 'not', default: 'not',
    enum: [366,181],
    '@': 367,
    '': 0
  },
  /* 365 */ { '@': [365,777], '': 0 },
  /* 366 */ { not: [0,442], null: 'not', default: 'not', '': 0 },
  /* 367 */ { '@': [367,777], '': 368 },
  /* 368 */ { enum: [369,181], '': 0 },
  /* 369 */ { not: [0,442], null: 'not', default: 'not', '': 0 },
  /* 370 */ [371,409],
  /* 371 */ { not: [372,442], null: 'not', default: 'not', '': 372 },
  /* 372 */ ['g',373],
  /* 373 */ { '@': [373,777], '': 0 },
  /* 374 */ { '[': [375,430], '': 375 },
  /* 375 */ ['mk',376,'to'],
  /* 376 */ { one: ['ck',377,1], many: 'one', '': 377 },
  /* 377 */ [378,40],
  /* 378 */ {
    on: ['ck',379],
    '{': [380,196],
    '': 380
  },
  /* 379 */ [381,642],
  /* 380 */ { not: [381,442], null: 'not', default: 'not', '': 381 },
  /* 381 */ ['g',382],
  /* 382 */ { '@': [382,777], '': 0 },
  /* 383 */ { '[': [384,430], '': 384 },
  /* 384 */ ['mk',385,'of'],
  /* 385 */ { one: ['ck',386,1], many: 'one', '': 386 },
  /* 386 */ {
    Id: [387,40],
    '{': 392,
  },
  /* 387 */ {
    on: ['ck',388],
    '{': [389,196],
    '': 389
  },
  /* 388 */ [390,642],
  /* 389 */ { not: [390,442], null: 'not', default: 'not', '': 390 },
  /* 390 */ ['g',391],
  /* 391 */ { '@': [391,777], '': 0 },
  /* 392 */ [0,160],
  /* 393 */ ['mk',394,'of'],
  /* 394 */ {
    '{': [395,160],
    Id: [396,409],
    type: [396,401,1],
  },
  /* 395 */ { not: [0,448], null: 'not', '': 0 },
  /* 396 */ { not: [397,448], null: 'not', '': 397 },
  /* 397 */ { enum: [398,181], '': 399 },
  /* 398 */ { not: [0,448], null: 'not', '': 0 },
  /* 399 */ { '@': [399,777], '': 0 },
  'typeTypeOf',
  /* 401 */ ['mk',402,'type'],
  /* 402 */ ['mk',403,'of'],
  /* 403 */ [404,40],
  /* 404 */ { ':': ['c',405], '': 0 },
  /* 405 */ ['miA',406],
  /* 406 */ { '.': ['c',407], '': 0 },
  /* 407 */ ['miA',406],
  'typeRefOptArgs',
  /* 409 */ [410,40],
  /* 410 */ {
    ':': ['c',411],
    '(': ['c',414],
    '': 0
  },
  /* 411 */ ['miA',412],
  /* 412 */ { '.': ['c',413], '': 0 },
  /* 413 */ ['miA',412],
  /* 414 */ {
    Number: ['c',415],
    Id: 417,
  },
  /* 415 */ { ',': ['c',416], '': 419 },
  /* 416 */ {
    Number: ['c',415],
    floating: ['ck',415], variable: 'floating',
    ')': 419,
  },
  /* 417 */ { Id: [418,426], '': 419 },
  /* 418 */ {
    ',': ['c',417],
    ')': 419,
  },
  /* 419 */ ['m',0,')'],
  'typeNamedArgsList',
  /* 421 */ ['m',422,'('],
  /* 422 */ { Id: [423,426], '': 424 },
  /* 423 */ {
    ',': ['c',422],
    ')': 424,
  },
  /* 424 */ ['m',0,')'],
  'typeNamedArg',
  /* 426 */ ['mi',427],
  /* 427 */ ['m',428,':'],
  /* 428 */ {
    Number: ['c',0],
    floating: ['ck',0], variable: 'floating',
  },
  'cardinality',
  /* 430 */ ['m',431,'['],
  /* 431 */ {
    '*': ['c',432],
    Number: ['c',434],
    '': 436
  },
  /* 432 */ { ',': ['c',433], '': 436 },
  /* 433 */ [436,438],
  /* 434 */ {
    ',': ['c',435],
    '..': [436,439],
    '': 436
  },
  /* 435 */ [436,438],
  /* 436 */ ['m',0,']'],
  'targetCardinality',
  /* 438 */ {
    '*': ['c',0],
    Number: ['c',439],
  },
  /* 439 */ { '..': ['c',440], '': 0 },
  /* 440 */ {
    '*': ['c',0],
    Number: ['c',0],
  },
  'nullabilityAndDefault',
  /* 442 */ {
    not: [443,448], null: 'not',
    default: ['ck',445],
  },
  /* 443 */ { default: ['ck',444], '': 0 },
  /* 444 */ [0,650],
  /* 445 */ [446,650],
  /* 446 */ { not: [0,448], null: 'not', '': 0 },
  'nullability',
  /* 448 */ {
    null: ['ck',0],
    not: ['ck',449],
  },
  /* 449 */ ['mk',0,'null'],
  'queryEOF',
  /* 451 */ [452,461],
  /* 452 */ { ';': ['c',453], '': 453 },
  /* 453 */ ['m',0,'EOF'],
  'projectionSpec',
  /* 455 */ ['mk',456,'projection'],
  /* 456 */ ['mk',457,'on'],
  /* 457 */ [458,520],
  /* 458 */ { '{': [459,566], '': 459 },
  /* 459 */ { excluding: [0,559], '': 0 },
  'queryExpression',
  /* 461 */ {
    '(': ['c',462],
    Id: [464,471],
  },
  /* 462 */ [463,461],
  /* 463 */ ['m',464,')'],
  /* 464 */ {
    intersect: ['ck',465],
    minus: ['ck',465], except: 'minus',
    union: ['ck',466],
    '': 468
  },
  /* 465 */ { distinct: ['ck',467], '': 467 },
  /* 466 */ { all: ['ck',467], distinct: 'all', '': 467 },
  /* 467 */ [464,461],
  /* 468 */ { limit: ['g',469], order: 'limit', '': 0 },
  /* 469 */ [0,622],
  'selectQuery',
  /* 471 */ ['mk',472,'select'],
  /* 472 */ {
    from: ['ck',473],
    all: ['ck',483,1], distinct: 'all',
    Id: 483, '#': 483, '(': 483, '*': 483, '+': 483, '-': 483, ':': 483, '?': 483, '@': 483, '{': 483, key: 483, not: 483, case: 483, cast: 483, null: 483, true: 483, false: 483, Number: 483, String: 483, exists: 483, QuotedLiteral: 483,
  },
  /* 473 */ [474,490],
  /* 474 */ { mixin: ['ck',475], '': 480 },
  /* 475 */ ['m',476,'{'],
  /* 476 */ { Id: [477,206], '': 478 },
  /* 477 */ {
    ';': ['c',476],
    '}': 478,
  },
  /* 478 */ ['m',479,'}'],
  /* 479 */ ['mk',480,'into'],
  /* 480 */ { all: ['ck',481], distinct: 'all', '': 481 },
  /* 481 */ { '{': [482,566], '': 482 },
  /* 482 */ { excluding: [488,559], '': 488 },
  /* 483 */ {
    '*': ['c',484],
    Id: [484,576], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', '@': 'Id', '{': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
  },
  /* 484 */ { ',': ['c',485], '': 486 },
  /* 485 */ {
    '*': ['c',484],
    Id: [484,576], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', '@': 'Id', '{': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
  },
  /* 486 */ ['mk',487,'from'],
  /* 487 */ [488,490],
  /* 488 */ { group: [0,610], where: 'group', having: 'group', '': 0 },
  'querySource',
  /* 490 */ [491,495],
  /* 491 */ { ',': 492, '': 0 },
  /* 492 */ { ',': ['c',493], '': 0 },
  /* 493 */ [492,495],
  'tableExpression',
  /* 495 */ {
    '(': [496,507],
    Id: [497,520],
  },
  /* 496 */ ['g',497],
  /* 497 */ {
    cross: ['ck',498],
    inner: ['ck',501],
    full: ['ck',500], left: 'full', right: 'full',
    join: 502,
    '': 0
  },
  /* 498 */ ['mk',499,'join'],
  /* 499 */ {
    '(': [497,507],
    Id: [497,520],
  },
  /* 500 */ { outer: ['ck',501], '': 501 },
  /* 501 */ { one: [502,514], many: 'one', exact: 'one', '': 502 },
  /* 502 */ ['mk',503,'join'],
  /* 503 */ [504,495],
  /* 504 */ ['mk',505,'on'],
  /* 505 */ [497,642],
  'tableOrQueryParens',
  /* 507 */ ['m',508,'('],
  /* 508 */ {
    '(': [509,507],
    Id: [510,495],
    select: [510,461],
  },
  /* 509 */ {
    full: [510,496], join: 'full', left: 'full', cross: 'full', inner: 'full', right: 'full',
    limit: [510,464], minus: 'limit', order: 'limit', union: 'limit', except: 'limit', intersect: 'limit',
    '': 510
  },
  /* 510 */ ['m',511,')'],
  /* 511 */ {
    as: ['ck',512],
    Id: ['ci',0],
    '': 0
  },
  /* 512 */ ['mi',0],
  'joinCardinality',
  /* 514 */ {
    exact: ['ck',515],
    one: 515,
    many: ['ck',516],
  },
  /* 515 */ ['mk',516,'one'],
  /* 516 */ ['mk',517,'to'],
  /* 517 */ {
    exact: ['ck',518],
    one: 518,
    many: ['ck',0],
  },
  /* 518 */ ['mk',0,'one'],
  'fromRefWithOptAlias',
  /* 520 */ [521,527],
  /* 521 */ { ':': ['c',522], '': 523 },
  /* 522 */ [523,527],
  /* 523 */ {
    as: ['ck',524],
    Id: ['ci',0],
    '': 525
  },
  /* 524 */ ['mi',0],
  /* 525 */ ['g',0,[]],
  'fromPath',
  /* 527 */ ['mi',528],
  /* 528 */ { '(': [529,533], '[': '(', '': 529 },
  /* 529 */ { '.': ['c',530], '': 0 },
  /* 530 */ ['miA',531],
  /* 531 */ { '(': [529,533], '[': '(', '': 529 },
  'fromArgumentsAndFilter',
  /* 533 */ { '(': ['c',534], '': 538 },
  /* 534 */ { Id: 535 },
  /* 535 */ { Id: [536,540], '': 537 },
  /* 536 */ {
    ',': ['c',535],
    ')': 537,
  },
  /* 537 */ ['m',538,')'],
  /* 538 */ { '[': [0,544], '': 0 },
  'fromNamedArgument',
  /* 540 */ ['mi',541],
  /* 541 */ ['m',542,':'],
  /* 542 */ [0,650],
  'cardinalityAndFilter',
  /* 544 */ ['m',545,'['],
  /* 545 */ { Number: ['c',546], '': 547 },
  /* 546 */ ['m',547,':'],
  /* 547 */ [548,550],
  /* 548 */ ['m',0,']'],
  'filterClauses',
  /* 550 */ {
    where: ['ck',551],
    Id: 551, '#': 551, '(': 551, '+': 551, '-': 551, ':': 551, '?': 551, not: 551, case: 551, cast: 551, null: 551, true: 551, false: 551, Number: 551, String: 551, exists: 551, QuotedLiteral: 551,
    group: ['g',552,1], limit: 'group', order: 'group', having: 'group',
    '': 552
  },
  /* 551 */ [552,642],
  /* 552 */ { group: 553, '': 554 },
  /* 553 */ [554,616],
  /* 554 */ { having: ['ck',555], '': 556 },
  /* 555 */ [556,642],
  /* 556 */ { limit: 557, order: 557, '': 0 },
  /* 557 */ [0,622],
  'excludingClause',
  /* 559 */ ['mk',560,'excluding'],
  /* 560 */ ['m',561,'{'],
  /* 561 */ { Id: 562 },
  /* 562 */ { Id: ['ciA',563], '': 564 },
  /* 563 */ {
    ',': ['c',562],
    '}': 564,
  },
  /* 564 */ ['m',0,'}'],
  'selectItemsList',
  /* 566 */ ['m',567,'{'],
  /* 567 */ {
    '*': ['c',568],
    Id: [568,576], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', '@': 'Id', '{': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    '': 569
  },
  /* 568 */ {
    ',': ['c',567],
    '}': 569,
  },
  /* 569 */ ['m',0,'}'],
  'nestedSelectItemsList',
  /* 571 */ ['m',572,'{'],
  /* 572 */ {
    '*': ['c',573],
    Id: [573,576], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', '@': 'Id', '{': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    '': 574
  },
  /* 573 */ {
    ',': ['c',572],
    '}': 574,
  },
  /* 574 */ ['m',0,'}'],
  'selectItemDef',
  /* 576 */ { '@': [576,780], '': 577 },
  /* 577 */ { virtual: ['ck',578,1], '': 578 },
  /* 578 */ { key: ['ck',579], '': 579 },
  /* 579 */ {
    Id: [580,650], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    '{': [587,571],
  },
  /* 580 */ {
    as: ['ck',581],
    Id: ['ci',582],
    '': 582
  },
  /* 581 */ ['mi',582],
  /* 582 */ {
    '{': 583,
    '.': ['c',585],
    '': 590
  },
  /* 583 */ [584,571],
  /* 584 */ { excluding: [590,559], '': 590 },
  /* 585 */ {
    '{': [586,571],
    '*': ['c',590],
  },
  /* 586 */ { excluding: [590,559], '': 590 },
  /* 587 */ { excluding: [588,559], '': 588 },
  /* 588 */ ['mk',589,'as'],
  /* 589 */ ['mi',590],
  /* 590 */ ['g',591],
  /* 591 */ { '@': [591,783], '': 592 },
  /* 592 */ { ':': ['c',593], '': 0 },
  /* 593 */ {
    type: [607,401,1],
    localized: ['ck',594,1],
    Id: 594,
    redirected: ['ck',595,1],
    association: ['ck',599,1],
    composition: ['ck',601,1],
  },
  /* 594 */ [607,409],
  /* 595 */ ['mk',596,'to'],
  /* 596 */ [597,40],
  /* 597 */ {
    on: ['ck',598],
    '{': [607,196],
    '': 607
  },
  /* 598 */ [607,642],
  /* 599 */ { '[': [600,430], '': 600 },
  /* 600 */ ['mk',603,'to'],
  /* 601 */ { '[': [602,430], '': 602 },
  /* 602 */ ['mk',603,'of'],
  /* 603 */ { one: ['ck',604,1], many: 'one', '': 604 },
  /* 604 */ [605,40],
  /* 605 */ ['mk',606,'on'],
  /* 606 */ [607,642],
  /* 607 */ ['g',608],
  /* 608 */ { '@': [608,777], '': 0 },
  'whereGroupByHaving',
  /* 610 */ { where: ['ck',611], '': 612 },
  /* 611 */ [612,642],
  /* 612 */ { group: [613,616], '': 613 },
  /* 613 */ { having: ['ck',614], '': 0 },
  /* 614 */ [0,642],
  'groupByClause',
  /* 616 */ ['mk',617,'group'],
  /* 617 */ ['mk',618,'by'],
  /* 618 */ [619,650],
  /* 619 */ { ',': ['c',620], '': 0 },
  /* 620 */ [619,650],
  'orderByLimitOffset',
  /* 622 */ { order: [623,628], '': 623 },
  /* 623 */ { limit: ['ck',624], '': 0 },
  /* 624 */ [625,650],
  /* 625 */ { offset: ['ck',626], '': 0 },
  /* 626 */ [0,650],
  'orderByClause',
  /* 628 */ ['mk',629,'order'],
  /* 629 */ ['mk',630,'by'],
  /* 630 */ [631,634],
  /* 631 */ { ',': ['c',632], '': 0 },
  /* 632 */ [631,634],
  'orderByExpression',
  /* 634 */ [635,650],
  /* 635 */ { asc: ['ck',636], desc: 'asc', '': 636 },
  /* 636 */ { nulls: ['ck',637], '': 0 },
  /* 637 */ { last: ['ck',0], first: 'last' },
  'conditionEOF',
  /* 639 */ [640,650],
  /* 640 */ ['m',0,'EOF'],
  'condition',
  /* 642 */ [0,650],
  'valuePath',
  /* 644 */ ['mi',645],
  /* 645 */ { '(': [646,711], '[': '(', '': 646 },
  /* 646 */ { '.': ['c',647], '': 0 },
  /* 647 */ ['miA',648],
  /* 648 */ { '(': [646,711], '[': '(', '': 646 },
  'expression',
  /* 650 */ {
    '(': [651,678],
    '#': [661,774], null: '#', true: '#', false: '#', Number: '#', String: '#', QuotedLiteral: '#',
    ':': ['c',652],
    '?': ['c',661],
    Id: [654,644],
    new: [661,690,1],
    exists: ['ck',657],
    case: [661,693],
    cast: [661,704],
    '+': ['c',660], '-': '+',
    not: ['ck',660],
  },
  /* 651 */ ['g',661],
  /* 652 */ {
    Id: ['ciA',653],
    Number: ['c',661],
  },
  /* 653 */ { '.': [661,646], '': 661 },
  /* 654 */ { over: ['ck',655,1], '': 656 },
  /* 655 */ [656,738],
  /* 656 */ ['g',661],
  /* 657 */ {
    '(': ['c',658],
    Id: [661,644],
    '?': ['c',661],
  },
  /* 658 */ [659,461],
  /* 659 */ ['m',661,')'],
  /* 660 */ [661,650],
  /* 661 */ {
    '*': ['c',665], '/': '*',
    '+': ['c',665], '-': '+',
    '||': ['c',665],
    and: ['ck',665,1],
    or: ['ck',665,1],
    '?': ['c',662],
    '<': ['c',664], '=': '<', '>': '<', '!=': '<', '<=': '<', '<>': '<', '>=': '<',
    is: ['ck',666,1],
    not: ['ck',668],
    in: ['g',668,1], like: 'in', between: 'in',
    '': 0
  },
  /* 662 */ [663,650],
  /* 663 */ ['m',665,':'],
  /* 664 */ { all: ['ck',665], any: 'all', some: 'all', '': 665 },
  /* 665 */ [676,650],
  /* 666 */ { not: ['ck',667], '': 667 },
  /* 667 */ ['mk',676,'null'],
  /* 668 */ {
    between: ['ck',669],
    in: ['ck',672],
    like: ['ck',673],
  },
  /* 669 */ [670,650],
  /* 670 */ ['mk',671,'and'],
  /* 671 */ [676,650],
  /* 672 */ [676,650],
  /* 673 */ [674,650],
  /* 674 */ { escape: ['ck',675,1], '': 676 },
  /* 675 */ [676,650],
  /* 676 */ ['g',661],
  'expressionOrQueryParens',
  /* 678 */ ['m',679,'('],
  /* 679 */ {
    '(': [680,678],
    Id: [682,650], '#': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', not: 'Id', case: 'Id', cast: 'Id', null: 'Id', true: 'Id', false: 'Id', Number: 'Id', String: 'Id', exists: 'Id', QuotedLiteral: 'Id',
    select: [683,461],
  },
  /* 680 */ {
    '*': [681,651], '+': '*', '-': '*', '/': '*', '<': '*', '=': '*', '>': '*', '?': '*', '!=': '*', '<=': '*', '<>': '*', '>=': '*', in: '*', is: '*', or: '*', '||': '*', and: '*', not: '*', like: '*', between: '*',
    ',': [683,685],
    limit: [683,464], minus: 'limit', order: 'limit', union: 'limit', except: 'limit', intersect: 'limit',
    '': 683
  },
  /* 681 */ { ',': [683,685], '': 683 },
  /* 682 */ { ',': [683,685], '': 683 },
  /* 683 */ ['m',0,')'],
  'continueExpressionslist',
  /* 685 */ ['m',686,','],
  /* 686 */ { Id: 687, '#': 687, '(': 687, '+': 687, '-': 687, ':': 687, '?': 687, not: 687, case: 687, cast: 687, null: 687, true: 687, false: 687, Number: 687, String: 687, exists: 687, QuotedLiteral: 687 },
  /* 687 */ { Id: [688,650], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', not: 'Id', case: 'Id', cast: 'Id', null: 'Id', true: 'Id', false: 'Id', Number: 'Id', String: 'Id', exists: 'Id', QuotedLiteral: 'Id', '': 0 },
  /* 688 */ { ',': ['c',687], '': 0 },
  'newAndValuePath',
  /* 690 */ ['mk',691,'new'],
  /* 691 */ [0,644],
  'caseExpression',
  /* 693 */ ['mk',694,'case'],
  /* 694 */ {
    Id: [695,650], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', not: 'Id', case: 'Id', cast: 'Id', null: 'Id', true: 'Id', false: 'Id', Number: 'Id', String: 'Id', exists: 'Id', QuotedLiteral: 'Id',
    when: ['g',695,1],
    '': 695
  },
  /* 695 */ { when: 696 },
  /* 696 */ { when: ['ck',697], '': 700 },
  /* 697 */ [698,650],
  /* 698 */ ['mk',699,'then'],
  /* 699 */ [696,650],
  /* 700 */ { else: ['ck',701], '': 702 },
  /* 701 */ [702,650],
  /* 702 */ ['mk',0,'end'],
  'castFunction',
  /* 704 */ ['mk',705,'cast'],
  /* 705 */ ['m',706,'('],
  /* 706 */ [707,650],
  /* 707 */ ['mk',708,'as'],
  /* 708 */ [709,409],
  /* 709 */ ['m',0,')'],
  'argumentsAndFilter',
  /* 711 */ { '(': ['c',712], '': 731 },
  /* 712 */ { Id: ['ciA',719], '': 713 },
  /* 713 */ { Id: [714,733], '#': 'Id', '(': 'Id', '*': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', Id_all: 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id', '': 730 },
  /* 714 */ { ',': ['c',715], '': 716 },
  /* 715 */ {
    Id: [714,733], '#': 'Id', '(': 'Id', '*': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', not: 'Id', case: 'Id', cast: 'Id', null: 'Id', true: 'Id', false: 'Id', Id_all: 'Id', Number: 'Id', String: 'Id', exists: 'Id', QuotedLiteral: 'Id',
    ')': ['g',716,1], order: ')',
  },
  /* 716 */ { order: ['ck',717], '': 730 },
  /* 717 */ ['mk',718,'by'],
  /* 718 */ [730,753],
  /* 719 */ {
    ':': ['c',720],
    '=>': ['c',725],
  },
  /* 720 */ [721,650],
  /* 721 */ { ',': ['c',722], '': 730 },
  /* 722 */ {
    Id: ['ciA',723],
    ')': 730,
  },
  /* 723 */ ['m',724,':'],
  /* 724 */ [721,650],
  /* 725 */ [726,650],
  /* 726 */ { ',': ['c',727], '': 730 },
  /* 727 */ {
    Id: ['ciA',728],
    ')': 730,
  },
  /* 728 */ ['m',729,'=>'],
  /* 729 */ [726,650],
  /* 730 */ ['m',731,')'],
  /* 731 */ { '[': [0,544], '': 0 },
  'funcExpression',
  /* 733 */ {
    ' lookahead': 'lGenericIntroOrExpr',
    Id: [735,650], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    GenericExpr: ['ckA',735],
    GenericIntro: ['ckA',734],
  },
  /* 734 */ [735,650],
  /* 735 */ {
    ' lookahead': 'lGenericSeparator',
    GenericSeparator: ['ckA',736],
    '': 0
  },
  /* 736 */ {
    ' lookahead': 'lGenericExpr',
    Id: [735,650], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', ':': 'Id', '?': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    GenericExpr: ['ckA',735],
  },
  'overClause',
  /* 738 */ ['m',739,'('],
  /* 739 */ { partition: ['ck',740], '': 742 },
  /* 740 */ ['mk',741,'by'],
  /* 741 */ [742,749],
  /* 742 */ { order: ['ck',743], '': 745 },
  /* 743 */ ['mk',744,'by'],
  /* 744 */ [745,753],
  /* 745 */ { rows: ['ck',746], '': 747 },
  /* 746 */ [747,762],
  /* 747 */ ['m',0,')'],
  'expressionsAsXpr',
  /* 749 */ [750,650],
  /* 750 */ { ',': ['c',751], '': 0 },
  /* 751 */ [750,650],
  'orderByClauseAsXpr',
  /* 753 */ [754,757],
  /* 754 */ { ',': ['c',755], '': 0 },
  /* 755 */ [754,757],
  'orderBySpecAsXpr',
  /* 757 */ [758,650],
  /* 758 */ { asc: ['ck',759], desc: 'asc', '': 759 },
  /* 759 */ { nulls: ['ck',760], '': 0 },
  /* 760 */ { last: ['ck',0], first: 'last' },
  'windowFrameClause',
  /* 762 */ {
    unbounded: ['ck',763],
    Number: ['c',763],
    current: ['ck',764],
    between: ['ck',765],
  },
  /* 763 */ ['mk',0,'preceding'],
  /* 764 */ ['mk',0,'row'],
  /* 765 */ [766,769],
  /* 766 */ ['mk',767,'and'],
  /* 767 */ [0,769],
  'windowFrameBoundSpec',
  /* 769 */ {
    unbounded: ['ck',770],
    Number: ['c',770],
    current: ['ck',772],
  },
  /* 770 */ {
    following: ['ck',771],
    preceding: ['ck',771],
  },
  /* 771 */ ['g',0,[]],
  /* 772 */ ['mk',0,'row'],
  'literalValue',
  /* 774 */ {
    '#': ['c',775],
    null: ['ck',0],
    true: ['ck',0], false: 'true',
    Number: ['c',0],
    String: ['c',0],
    QuotedLiteral: ['c',0],
  },
  /* 775 */ ['mi',0],
  'annoAssignStd',
  /* 777 */ ['m',778,'@'],
  /* 778 */ {
    '(': [0,786],
    Id: [0,791],
  },
  'annoAssignCol',
  /* 780 */ ['m',781,'@'],
  /* 781 */ {
    '(': [0,786],
    Id: [0,791],
  },
  'annoAssignMid',
  /* 783 */ ['m',784,'@'],
  /* 784 */ {
    '(': [0,786],
    Id: [0,795],
  },
  'annoAssignParen',
  /* 786 */ ['m',787,'('],
  /* 787 */ { Id: [788,791], '': 789 },
  /* 788 */ {
    ',': ['c',787],
    ')': 789,
  },
  /* 789 */ ['m',0,')'],
  'annoAssignBase',
  /* 791 */ [792,795],
  /* 792 */ { ':': ['c',793], '': 0 },
  /* 793 */ [0,817],
  'annoNamePath',
  /* 795 */ ['miA',796],
  /* 796 */ { '.': ['c',797], '': 799 },
  /* 797 */ {
    Id: ['ciA',796],
    '@': ['c',798],
  },
  /* 798 */ ['miA',796],
  /* 799 */ { '#': [0,808], '': 0 },
  'annoPath',
  /* 801 */ {
    Id: ['ciA',803],
    '@': ['c',802],
  },
  /* 802 */ ['miA',803],
  /* 803 */ { '.': ['c',804], '': 806 },
  /* 804 */ {
    Id: ['ciA',803],
    '@': ['c',805],
  },
  /* 805 */ ['miA',803],
  /* 806 */ { '#': [0,808], '': 0 },
  'annoPathVariant',
  /* 808 */ ['m',809,'#'],
  /* 809 */ ['miA',810],
  /* 810 */ { '.': ['c',811], '': 0 },
  /* 811 */ ['miA',810],
  'annoStructValue',
  /* 813 */ [814,801],
  /* 814 */ { ':': ['c',815], '': 0 },
  /* 815 */ [0,817],
  'annoValue',
  /* 817 */ {
    '#': [0,774], null: '#', true: '#', false: '#', Number: '#', String: '#', QuotedLiteral: '#',
    '+': ['c',818], '-': '+',
    Id: [0,801], '@': 'Id',
    '{': ['c',819],
    '[': ['c',822],
    '(': ['c',829],
  },
  /* 818 */ ['m',0,'Number'],
  /* 819 */ { Id: [820,813], '@': 'Id', '': 821 },
  /* 820 */ {
    ',': ['c',819],
    '}': 821,
  },
  /* 821 */ ['m',0,'}'],
  /* 822 */ {
    Id: [827,817], '#': 'Id', '(': 'Id', '+': 'Id', '-': 'Id', '@': 'Id', '[': 'Id', '{': 'Id', Number: 'Id', String: 'Id', QuotedLiteral: 'Id',
    '...': ['c',823],
    '': 828
  },
  /* 823 */ { up: ['ck',824], '': 826 },
  /* 824 */ ['mk',825,'to'],
  /* 825 */ [826,817],
  /* 826 */ ['g',827],
  /* 827 */ {
    ',': ['c',822],
    ']': 828,
  },
  /* 828 */ ['m',0,']'],
  /* 829 */ [830,642],
  /* 830 */ ['m',0,')'],
];
class CdlParser extends AstBuildingParser {
  constructor(lexer,...args) { super(lexer,keywords,table,...args) }
  start($,$next) {
    $.source??=new XsnSource( 'cdl' );
    this.rule_(1,$next);
    for(;;) switch(this.s) {
      case 1: switch(this.lk()) {
        case 'namespace': this.gc(3,'namespaceRestriction') && this.namespaceDeclaration({source:$.source},2); continue;
        case 'using': this.usingDeclaration({source:$.source},2); continue;
        case '@': case 'type': case 'view': case 'event': case 'action': case 'aspect': case 'define': case 'entity': case 'extend': case 'context': case 'service': case 'abstract': case 'annotate': case 'function': case 'annotation': if(this.artifactDefOrExtend({outer:$.source},2)) {this.namespaceRestriction();} continue;
        default: this.s=3; continue;
      }
      case 2: switch(this.l()) {
        case ';': this.c(1); continue;
        case 'EOF': this.s=3; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(1); continue;
        default: this.e(); continue;
      }
      case 3: if(this.m(0,'EOF')) { this.docComment( null ); } continue;
      default: return this.exit_();
    }
  }
  artifactsBlock($,$next) {
    this.rule_(5,$next);
    for(;;) switch(this.s) {
      case 5: if(this.m(6,'{')) {this.vocabularyRestriction(); $.art.artifacts = this.createDict( $.start ); $.art.extensions = []; } continue;
      case 6: switch(this.l()) {
        case 'Id': case '@': this.artifactDefOrExtend({outer:$.art},7); continue;
        default: this.s=8; continue;
      }
      case 7: switch(this.l()) {
        case ';': this.c(6); continue;
        case '}': this.s=8; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(6); continue;
        default: this.e(); continue;
      }
      case 8: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.art.artifacts ); } continue;
      default: return this.exit_();
    }
  }
  artifactDefOrExtend($,$next) {
    this.rule_(10,$next);
    let art=new XsnArtifact();
    { art.location = this.startLocation();
this.docComment( art ); }
    for(;;) switch(this.s) {
      case 10: switch(this.l()) {
        case '@': this.annoAssignStd({art},10); continue;
        default: this.s=11; continue;
      }
      case 11: switch(this.lk()) {
        case 'define': this.ck(12); continue;
        case 'type': case 'view': case 'event': case 'action': case 'aspect': case 'entity': case 'context': case 'service': case 'abstract': case 'function': case 'annotation': this.s=12; continue;
        case 'extend': if(this.ec('extensionRestriction') && this.ck(13)) { art.kind = 'extend'; } continue;
        case 'annotate': this.ec('extensionRestriction') && this.ck(14); continue;
        default: this.e(); continue;
      }
      case 12: switch(this.lk()) {
        case 'service': this.serviceDef({art,outer:$.outer},0); continue;
        case 'context': this.contextDef({art,outer:$.outer},0); continue;
        case 'annotation': this.ec('vocabularyRestriction') && this.annotationDef({art,outer:$.outer},0); continue;
        case 'type': this.typeDef({art,outer:$.outer},0); continue;
        case 'aspect': case 'abstract': this.aspectDef({art,outer:$.outer},0); continue;
        case 'entity': this.entityDef({art,outer:$.outer},0); continue;
        case 'view': this.viewDef({art,outer:$.outer},0); continue;
        case 'event': this.eventDef({art,outer:$.outer},0); continue;
        case 'action': this.actionMainDef({art,outer:$.outer},0); continue;
        case 'function': this.functionMainDef({art,outer:$.outer},0); continue;
        default: this.e(); continue;
      }
      case 13: switch(this.lk()) {
        case 'Id': this.extendArtifact({art,outer:$.outer},0); continue;
        case 'service': this.lP() && this.extendService({art,outer:$.outer},0); continue;
        case 'context': this.lP() && this.extendContext({art,outer:$.outer},0); continue;
        case 'type': this.lP() && this.extendType({art,outer:$.outer},0); continue;
        case 'aspect': case 'entity': this.lP() && this.extendEntityOrAspect({art,outer:$.outer},0); continue;
        case 'projection': this.lP() && this.extendProjection({art,outer:$.outer},0); continue;
        default: this.ei(); continue;
      }
      case 14: this.annotateArtifact({art,outer:$.outer},0); continue;
      default: return this.exit_();
    }
  }
  namespaceDeclaration($,$next) {
    this.rule_(16,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 16: this.mk(17,'namespace'); continue;
      case 17: if(this.namePath(_={category:'Namespace'},0)) {name=_.name; $.source.namespace ??= { kind: 'namespace', name: name }; } continue;
      default:
      this.attachLocation( $.source.namespace );
      return this.exit_();
    }
  }
  usingDeclaration($,$next) {
    this.rule_(19,$next);
    let decl={ kind: 'using' };
    for(;;) switch(this.s) {
      case 19: this.mk(20,'using'); continue;
      case 20: switch(this.lk()) {
        case 'from': this.ck(21); continue;
        case 'Id': this.usingProxy({outer:$.source,proxy:decl},22); continue;
        case '{': this.s=24;{ $.source.usings.push( decl ); } continue;
        default: this.ei(); continue;
      }
      case 21: if(this.m(0,'String')) { $.source.dependencies.push( this.quotedLiteral() ); } continue;
      case 22: switch(this.lk()) {
        case 'from': this.ck(23); continue;
        default: this.gr([';']); continue;
      }
      case 23: if(this.m(0,'String')) { $.source.dependencies.push( decl.fileDep = this.quotedLiteral() ); } continue;
      case 24: if(this.m(25,'{')) { decl.usings = this.createArray(); } continue;
      case 25: switch(this.lk()) {
        case 'Id': this.s=26; continue;
        default: this.ei(); continue;
      }
      case 26: switch(this.l()) {
        case 'Id': this.usingProxy({outer:decl,proxy:{ kind: 'using' }},27); continue;
        default: this.s=28; continue;
      }
      case 27: switch(this.l()) {
        case ',': this.c(26); continue;
        case '}': this.s=28; continue;
        default: this.e(); continue;
      }
      case 28: if(this.m(29,'}')) {this.afterBrace(); this.finalizeDictOrArray( decl.usings ); } continue;
      case 29: switch(this.lk()) {
        case 'from': this.ck(30); continue;
        default: this.gr([';']); continue;
      }
      case 30: if(this.m(0,'String')) { $.source.dependencies.push( decl.fileDep = this.quotedLiteral() ); } continue;
      default:
      this.attachLocation( decl );
      return this.exit_();
    }
  }
  usingProxy($,$next) {
    this.rule_(32,$next);
    let extern; let _;
    for(;;) switch(this.s) {
      case 32: if(this.simplePath(_={category:'global'},33)) {extern=_.ref; $.proxy.extern = extern; $.outer.usings.push( $.proxy ); } continue;
      case 33: switch(this.lk()) {
        case 'as': this.ck(34); continue;
        default: if(this.gr([])) { this.classifyImplicitName( 'Using' ); } continue;
      }
      case 34: if(this.mi(0,'UsingAlias')) { $.proxy.name = this.identAst(); } continue;
      default:
      this.attachLocation( $.proxy );
      return this.exit_();
    }
  }
  namePath($,$next) {
    $.name??=new XsnName();
    this.rule_(36,$next);
    for(;;) switch(this.s) {
      case 36: if(this.mi(37,$.category)) { $.name.path = [ this.identAst() ]; } continue;
      case 37: switch(this.l()) {
        case '.': this.c(38); continue;
        default: this.gr(['@']); continue;
      }
      case 38: if(this.miA(37,$.category)) { $.name.path.push( this.identAst() ); } continue;
      default:
      this.attachLocation( $.name );
      return this.exit_();
    }
  }
  simplePath($,$next) {
    $.ref??={};
    this.rule_(40,$next);
    for(;;) switch(this.s) {
      case 40: if(this.mi(41,$.category)) { $.ref.path = [ this.identAst() ]; } continue;
      case 41: switch(this.l()) {
        case '.': this.c(42); continue;
        default: this.gr([]); continue;
      }
      case 42: if(this.miA(41,$.category)) { $.ref.path.push( this.identAst() ); } continue;
      default:
      this.attachLocation( $.ref );
      return this.exit_();
    }
  }
  serviceDef($,$next) {
    this.rule_(44,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 44: this.mk(45,'service'); continue;
      case 45: if(this.namePath(_={category:'Service'},46)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'service', name );
this.docComment( $.art ); } continue;
      case 46: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},46); continue;
        default: this.s=47; continue;
      }
      case 47: switch(this.l()) {
        case '{': this.artifactsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  contextDef($,$next) {
    this.rule_(49,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 49: this.mk(50,'context'); continue;
      case 50: if(this.namePath(_={category:'Context'},51)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'context', name );
this.docComment( $.art ); } continue;
      case 51: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},51); continue;
        default: this.s=52; continue;
      }
      case 52: switch(this.l()) {
        case '{': this.artifactsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  annotationDef($,$next) {
    this.rule_(54,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 54: this.mk(55,'annotation'); continue;
      case 55: if(this.namePath(_={category:'AnnoDef'},56)) {name=_.name; this.addDef( $.art, $.outer, 'vocabularies', 'annotation', name );
this.docComment( $.art ); } continue;
      case 56: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},56); continue;
        default: this.s=57; continue;
      }
      case 57: this.typeOrIncludesSpec({art:$.art},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  typeDef($,$next) {
    this.rule_(59,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 59: this.mk(60,'type'); continue;
      case 60: if(this.namePath(_={category:'Type'},61)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'type', name );
this.docComment( $.art ); } continue;
      case 61: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},61); continue;
        default: this.s=62; continue;
      }
      case 62: this.typeOrIncludesSpec({art:$.art},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  aspectDef($,$next) {
    this.rule_(64,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 64: switch(this.lk()) {
        case 'aspect': this.ck(66); continue;
        case 'abstract': if(this.ck(65)) { this.warning( 'syntax-deprecated-abstract', this.lb().location ); } continue;
        default: this.e(); continue;
      }
      case 65: this.mk(66,'entity'); continue;
      case 66: if(this.namePath(_={category:'Type'},67)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'aspect', name );
this.docComment( $.art ); } continue;
      case 67: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},67); continue;
        default: this.s=68; continue;
      }
      case 68: switch(this.l()) {
        case '{': this.elementsBlock({art:$.art},69); continue;
        case ':': this.c(70); continue;
        default: this.gr([';']); continue;
      }
      case 69: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 70: this.aspectColonSpec({art:$.art},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  aspectColonSpec($,$next) {
    this.rule_(72,$next);
    let incl; let _;
    for(;;) switch(this.s) {
      case 72: switch(this.l()) {
        case 'Id': if(this.simplePath(_={category:'artref'},73)) {incl=_.ref; $.art.includes ??= []; $.art.includes.push( incl ); } continue;
        default: this.s=74; continue;
      }
      case 73: switch(this.l()) {
        case ',': this.c(72); continue;
        default: this.s=74; continue;
      }
      case 74: switch(this.l()) {
        case '{': this.elementsBlock({art:$.art},75); continue;
        default: this.gr([';']); continue;
      }
      case 75: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default: return this.exit_();
    }
  }
  entityDef($,$next) {
    this.rule_(77,$next);
    let name; let incl; let query; let _;
    for(;;) switch(this.s) {
      case 77: this.mk(78,'entity'); continue;
      case 78: if(this.namePath(_={category:'Entity'},79)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'entity', name );
this.docComment( $.art ); } continue;
      case 79: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},79); continue;
        default: this.s=80; continue;
      }
      case 80: switch(this.l()) {
        case '(': this.paramsList({art:$.art},81); continue;
        default: this.s=81; continue;
      }
      case 81: switch(this.lk()) {
        case ':': if(this.c(82)) { $.art.includes ??= []; } continue;
        case '{': this.s=85; continue;
        case 'as': this.ck(86); continue;
        default: this.e(); continue;
      }
      case 82: switch(this.lk()) {
        case 'Id': this.s=83; continue;
        default: this.ei(); continue;
      }
      case 83: switch(this.l()) {
        case 'Id': if(this.simplePath(_={category:'artref'},84)) {incl=_.ref; $.art.includes.push( incl ); } continue;
        default: this.s=85; continue;
      }
      case 84: switch(this.l()) {
        case ',': this.c(83); continue;
        case '{': this.s=85; continue;
        default: this.e(); continue;
      }
      case 85: this.elementsBlock({art:$.art},89); continue;
      case 86: switch(this.lk()) {
        case '(': case 'select': if(this.queryExpression(_={},89)) {query=_.expr; $.art.query = query; $.art.$syntax = 'entity'; } continue;
        case 'projection': if(this.projectionSpec(_={},87)) {query=_.query; $.art.query = query; $.art.$syntax = 'projection'; } continue;
        default: this.e(); continue;
      }
      case 87: switch(this.lk()) {
        case 'group': case 'where': case 'having': this.whereGroupByHaving({query},88); continue;
        default: this.s=88; continue;
      }
      case 88: switch(this.lk()) {
        case 'limit': case 'order': this.orderByLimitOffset({query},89); continue;
        default: this.s=89; continue;
      }
      case 89: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  viewDef($,$next) {
    this.rule_(91,$next);
    let name; let query; let _;
    for(;;) switch(this.s) {
      case 91: this.mk(92,'view'); continue;
      case 92: if(this.namePath(_={category:'Entity'},93)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'entity', name );
this.docComment( $.art ); } continue;
      case 93: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},93); continue;
        default: this.s=94; continue;
      }
      case 94: switch(this.lk()) {
        case '(': this.paramsList({art:$.art},100); continue;
        case 'with': this.ck(95); continue;
        default: this.s=100; continue;
      }
      case 95: if(this.mk(96,'parameters')) { $.art.params = this.createDict(); } continue;
      case 96: this.paramDef({outer:$.art},97); continue;
      case 97: switch(this.l()) {
        case ',': this.c(98); continue;
        default: this.s=99; continue;
      }
      case 98: this.paramDef({outer:$.art},97); continue;
      case 99: this.s=100;{ this.finalizeDictOrArray( $.art.params ); } continue;
      case 100: this.mk(101,'as'); continue;
      case 101: if(this.queryExpression(_={},0)) {query=_.expr; $.art.query = query; $.art.$syntax = 'view'; } continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  eventDef($,$next) {
    this.rule_(103,$next);
    let name; let incl; let query; let _;
    for(;;) switch(this.s) {
      case 103: this.mk(104,'event'); continue;
      case 104: if(this.namePath(_={category:'Event'},105)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'event', name );
this.docComment( $.art ); } continue;
      case 105: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},105); continue;
        default: this.s=106; continue;
      }
      case 106: switch(this.l()) {
        case '{': this.elementsBlock({art:$.art},0); continue;
        case ':': this.c(107); continue;
        default: this.e(); continue;
      }
      case 107: switch(this.lk()) {
        case '{': this.elementsBlock({art:$.art},0); continue;
        case 'Id': if(this.simplePath(_={category:'artref'},108)) {incl=_.ref; $.art.type = incl; } continue;
        case 'projection': if(this.lP() && this.projectionSpec(_={},0)) {query=_.query; $.art.query = query; $.art.$syntax = 'projection'; } continue;
        default: this.ei(); continue;
      }
      case 108: switch(this.l()) {
        case ',': case '{': this.s=109;{ $.art.includes = [ $.art.type ]; delete $.art.type; } continue;
        default: this.s=113;{ this.docComment( $.art ); } continue;
      }
      case 109: switch(this.l()) {
        case ',': this.c(110); continue;
        default: this.s=112; continue;
      }
      case 110: switch(this.l()) {
        case 'Id': if(this.simplePath(_={category:'artref'},111)) {incl=_.ref; $.art.includes.push( incl ); } continue;
        default: this.s=112; continue;
      }
      case 111: switch(this.l()) {
        case ',': this.c(110); continue;
        case '{': this.s=112; continue;
        default: this.e(); continue;
      }
      case 112: this.elementsBlock({art:$.art},0); continue;
      case 113: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},113); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  actionMainDef($,$next) {
    this.rule_(115,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 115: this.mk(116,'action'); continue;
      case 116: if(this.namePath(_={category:'Action'},117)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'action', name );
this.docComment( $.art ); } continue;
      case 117: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},117); continue;
        default: this.s=118; continue;
      }
      case 118: this.paramsList({art:$.art},119); continue;
      case 119: switch(this.lk()) {
        case 'returns': this.returnsSpec({outer:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  functionMainDef($,$next) {
    this.rule_(121,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 121: this.mk(122,'function'); continue;
      case 122: if(this.namePath(_={category:'Action'},123)) {name=_.name; this.addDef( $.art, $.outer, 'artifacts', 'function', name );
this.docComment( $.art ); } continue;
      case 123: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},123); continue;
        default: this.s=124; continue;
      }
      case 124: this.paramsList({art:$.art},125); continue;
      case 125: this.returnsSpec({outer:$.art},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  actionsBlock($,$next) {
    this.rule_(127,$next);
    for(;;) switch(this.s) {
      case 127: if(this.mk(128,'actions')) { $.art.actions = this.createDict(); } continue;
      case 128: this.m(129,'{'); continue;
      case 129: switch(this.l()) {
        case 'Id': case '@': this.boundActionFunctionDef({outer:$.art},130); continue;
        default: this.s=131; continue;
      }
      case 130: switch(this.l()) {
        case ';': this.c(129); continue;
        case '}': this.s=131; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(129); continue;
        default: this.e(); continue;
      }
      case 131: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.art.actions ); } continue;
      default: return this.exit_();
    }
  }
  boundActionFunctionDef($,$next) {
    this.rule_(133,$next);
    let art=new XsnArtifact();
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 133: switch(this.l()) {
        case '@': this.annoAssignStd({art},133); continue;
        default: this.s=134; continue;
      }
      case 134: switch(this.lk()) {
        case 'action': this.ck(135); continue;
        case 'function': this.ck(139); continue;
        default: this.e(); continue;
      }
      case 135: if(this.mi(136,'BoundAction')) { this.addDef( art, $.outer, 'actions', 'action', this.identAst() );
this.docComment( art ); } continue;
      case 136: switch(this.l()) {
        case '@': this.annoAssignMid({art},136); continue;
        default: this.s=137; continue;
      }
      case 137: this.paramsList({art},138); continue;
      case 138: switch(this.lk()) {
        case 'returns': this.returnsSpec({outer:art},0); continue;
        default: this.gr([';','@','}','action','function']); continue;
      }
      case 139: if(this.mi(140,'BoundAction')) { this.addDef( art, $.outer, 'actions', 'function', this.identAst() );
this.docComment( art ); } continue;
      case 140: switch(this.l()) {
        case '@': this.annoAssignMid({art},140); continue;
        default: this.s=141; continue;
      }
      case 141: this.paramsList({art},142); continue;
      case 142: this.returnsSpec({outer:art},0); continue;
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  paramsList($,$next) {
    this.rule_(144,$next);
    for(;;) switch(this.s) {
      case 144: if(this.m(145,'(')) { $.art.params = this.createDict(); } continue;
      case 145: switch(this.l()) {
        case 'Id': case '@': this.paramDef({outer:$.art},146); continue;
        default: this.s=147; continue;
      }
      case 146: switch(this.l()) {
        case ',': this.c(145); continue;
        case ')': this.s=147; continue;
        default: this.e(); continue;
      }
      case 147: if(this.m(0,')')) { this.finalizeDictOrArray( $.art.params ); } continue;
      default: return this.exit_();
    }
  }
  paramDef($,$next) {
    this.rule_(149,$next);
    let art=new XsnArtifact();
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 149: switch(this.l()) {
        case '@': this.annoAssignStd({art},149); continue;
        default: this.s=150; continue;
      }
      case 150: if(this.miA(151,'Param')) { this.addDef( art, $.outer, 'params', 'param', this.identAst() );
this.docComment( art ); } continue;
      case 151: switch(this.l()) {
        case '@': this.annoAssignMid({art},151); continue;
        default: this.s=152; continue;
      }
      case 152: switch(this.l()) {
        case '{': this.elementsBlock({art},153); continue;
        case ':': this.c(154); continue;
        default: this.e(); continue;
      }
      case 153: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art},0); continue;
        default: this.gr([',']); continue;
      }
      case 154: this.typeExpression({art},0); continue;
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  returnsSpec($,$next) {
    this.rule_(156,$next);
    let art=new XsnArtifact();
    for(;;) switch(this.s) {
      case 156: if(this.mk(157,'returns')) {this.elementRestriction(); art.kind = 'param'; $.outer.returns = art;
this.docComment( art ); } continue;
      case 157: switch(this.l()) {
        case '@': this.annoAssignStd({art},157); continue;
        default: this.s=158; continue;
      }
      case 158: this.typeExpression({art},0); continue;
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  elementsBlock($,$next) {
    this.rule_(160,$next);
    for(;;) switch(this.s) {
      case 160: if(this.m(161,'{')) { $.art.elements = this.createDict(); } continue;
      case 161: switch(this.l()) {
        case 'Id': case '@': this.elementDef({outer:$.art,art:undefined},162); continue;
        default: this.s=163; continue;
      }
      case 162: switch(this.l()) {
        case ';': this.c(161); continue;
        case '}': this.s=163; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(161); continue;
        default: this.e(); continue;
      }
      case 163: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.art.elements ); } continue;
      default: return this.exit_();
    }
  }
  elementDef($,$next) {
    this.rule_(165,$next);
    let expr; let _;
    { $.art ??= new XsnArtifact();
this.docComment( $.art ); }
    for(;;) switch(this.s) {
      case 165: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},165); continue;
        default: this.s=166; continue;
      }
      case 166: switch(this.lk()) {
        case 'virtual': if(this.ckP(167,['Id','key'])) { $.art.virtual = this.valueWithLocation( true ); } continue;
        default: this.s=167; continue;
      }
      case 167: switch(this.lk()) {
        case 'key': if(this.ck(168)) { $.art.key = this.valueWithLocation( true ); } continue;
        default: this.s=168; continue;
      }
      case 168: switch(this.lk()) {
        case 'masked': if(this.ckP(169,['Id'])) { $.art.masked = this.valueWithLocation( true );
this.message( 'syntax-unsupported-masked', this.lb(), { keyword: 'masked' } ); } continue;
        default: this.s=169; continue;
      }
      case 169: switch(this.lk()) {
        case 'element': if(this.ckP(170,['Id'])) { $.art.$syntax = 'element'; } continue;
        default: this.s=170; continue;
      }
      case 170: if(this.mi(171,'Element')) { this.addDef( $.art, $.outer, 'elements', 'element', this.identAst() );
this.docComment( $.art ); } continue;
      case 171: switch(this.l()) {
        case '@': this.annoAssignMid({art:$.art},171); continue;
        default: this.s=172; continue;
      }
      case 172: switch(this.l()) {
        case '{': this.elementsBlock({art:$.art},173); continue;
        case ':': if(this.c(174)) {this.elementRestriction();} continue;
        default: this.s=175;{ this.elementRestriction(); } continue;
      }
      case 173: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},175); continue;
        default: this.s=175; continue;
      }
      case 174: this.typeExpression({art:$.art},175); continue;
      case 175: switch(this.l()) {
        case '=': this.gc(0,'calcOrDefaultRestriction') && this.c(176); continue;
        default: this.gr(['Id',';','@','}','key']); continue;
      }
      case 176: if(this.expression(_={},177)) {expr=_.expr; $.art.value = expr; } continue;
      case 177: switch(this.lk()) {
        case 'stored': if(this.ck(178)) { $.art.value.stored = this.valueWithLocation( true ); } continue;
        default: this.s=178; continue;
      }
      case 178: this.s=179;{ this.docComment( $.art ); } continue;
      case 179: switch(this.l()) {
        case '@': this.gc(0,'elementRestriction') && this.annoAssignStd({art:$.art},179); continue;
        default: this.gr(['Id',';','@','}','key']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  enumSymbolsBlock($,$next) {
    this.rule_(181,$next);
    for(;;) switch(this.s) {
      case 181: if(this.mk(182,'enum')) { $.art.enum = this.createDict(); } continue;
      case 182: this.m(183,'{'); continue;
      case 183: switch(this.l()) {
        case 'Id': case '@': this.enumSymbolDef({outer:$.art},184); continue;
        default: this.s=185; continue;
      }
      case 184: switch(this.l()) {
        case ';': this.c(183); continue;
        case '}': this.s=185; continue;
        default: this.e(); continue;
      }
      case 185: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.art.enum ); } continue;
      default: return this.exit_();
    }
  }
  enumSymbolDef($,$next) {
    this.rule_(187,$next);
    let art=new XsnArtifact();
    let sign; let value; let _;
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 187: switch(this.l()) {
        case '@': this.annoAssignStd({art},187); continue;
        default: this.s=188; continue;
      }
      case 188: if(this.mi(189,'Enum')) { this.addDef( art, $.outer, 'enum', 'enum', this.identAst() );
this.docComment( art ); } continue;
      case 189: switch(this.l()) {
        case '@': this.annoAssignStd({art},189); continue;
        default: this.s=190; continue;
      }
      case 190: switch(this.l()) {
        case '=': this.c(191); continue;
        default: this.gr([';','}']); continue;
      }
      case 191: switch(this.l()) {
        case 'String': if(this.c(193)) { art.value = this.quotedLiteral(); } continue;
        case 'Number': if(this.c(193)) { art.value = this.numberLiteral(); } continue;
        case '+': case '-': if(this.c(192)) sign=this.lb(); continue;
        case 'Id': case '#': case 'QuotedLiteral': if(this.literalValue(_={},193)) {value=_.expr; art.value = value; } continue;
        default: this.e(); continue;
      }
      case 192: if(this.m(193,'Number')) { art.value = this.numberLiteral( sign ); } continue;
      case 193: this.s=194;{ this.docComment( art ); } continue;
      case 194: switch(this.l()) {
        case '@': this.annoAssignStd({art},194); continue;
        default: this.gr([';','}']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  foreignKeysBlock($,$next) {
    this.rule_(196,$next);
    for(;;) switch(this.s) {
      case 196: if(this.m(197,'{')) { $.art.foreignKeys = this.createDict(); } continue;
      case 197: switch(this.l()) {
        case 'Id': case '@': this.foreignKeyDef({outer:$.art},198); continue;
        default: this.s=199; continue;
      }
      case 198: switch(this.l()) {
        case ',': this.c(197); continue;
        case '}': this.s=199; continue;
        default: this.e(); continue;
      }
      case 199: if(this.m(0,'}')) { this.finalizeDictOrArray( $.art.foreignKeys ); } continue;
      default: return this.exit_();
    }
  }
  foreignKeyDef($,$next) {
    this.rule_(201,$next);
    let art=new XsnArtifact(); let name;
    let ref; let _;
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 201: switch(this.l()) {
        case '@': this.annoAssignStd({art},201); continue;
        default: this.s=202; continue;
      }
      case 202: if(this.simplePath(_={category:'ref'},203)) {ref=_.ref; art.targetElement = ref; } continue;
      case 203: switch(this.lk()) {
        case 'as': this.ck(204); continue;
        default: if(this.gr([',','}'])) { this.classifyImplicitName( 'KeyImplicit', ref ); name = ref.path; } continue;
      }
      case 204: if(this.mi(0,'Key')) {name=this.lb(); name = this.identAst(); } continue;
      case 0: { this.addDef( art, $.outer, 'foreignKeys', 'key', name ); }
      default:
      this.attachLocation(art);
      return this.exit_();
    }
  }
  mixinElementDef($,$next) {
    this.rule_(206,$next);
    let art=new XsnArtifact();
    let assoc; let card; let target; let expr; let _;
    for(;;) switch(this.s) {
      case 206: if(this.mi(207,'Mixin')) { this.addDef( art, $.outer, 'mixin', 'mixin', this.identAst() ); } continue;
      case 207: this.m(208,':'); continue;
      case 208: switch(this.lk()) {
        case 'association': if(this.ck(209)) assoc=this.lb(); continue;
        case 'composition': if(this.ck(211)) assoc=this.lb(); continue;
        default: this.e(); continue;
      }
      case 209: switch(this.l()) {
        case '[': this.cardinality({art},210); continue;
        default: this.s=210; continue;
      }
      case 210: this.mk(213,'to'); continue;
      case 211: switch(this.l()) {
        case '[': this.cardinality({art},212); continue;
        default: this.s=212; continue;
      }
      case 212: this.mk(213,'of'); continue;
      case 213: switch(this.lk()) {
        case 'one': case 'many': if(this.ckP(214,['Id'])) card=this.lb(); continue;
        default: this.s=214; continue;
      }
      case 214: if(this.simplePath(_={category:'artref'},215)) {target=_.ref; this.setAssocAndComposition( art, assoc, card, target ); } continue;
      case 215: this.mk(216,'on'); continue;
      case 216: if(this.condition(_={},0)) {expr=_.expr; art.on = expr; } continue;
      default:
      this.attachLocation(art);
      return this.exit_();
    }
  }
  annotateArtifact($,$next) {
    this.rule_(218,$next);
    let name; let elemName; let _;
    for(;;) switch(this.s) {
      case 218: if(this.namePath(_={category:'Ext'},219)) name=_.name; continue;
      case 219: switch(this.lk()) {
        case ':': this.c(220); continue;
        case 'with': this.ck(225); continue;
        default: this.s=225; continue;
      }
      case 220: if(this.namePath(_={category:'ExtElement'},221)) {elemName=_.name; this.addExtension( $.art, $.outer, 'annotate', name, elemName.path ); } continue;
      case 221: switch(this.lk()) {
        case 'with': this.ck(222); continue;
        default: this.s=222; continue;
      }
      case 222: this.s=223;{ this.docComment( $.art ); } continue;
      case 223: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},223); continue;
        default: this.s=224; continue;
      }
      case 224: switch(this.l()) {
        case '{': this.annotateElementsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 225: this.s=226;{ this.addExtension( $.art, $.outer, 'annotate', name );
this.docComment( $.art ); } continue;
      case 226: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},226); continue;
        default: this.s=227; continue;
      }
      case 227: switch(this.l()) {
        case '(': this.annotateParamsBlock({art:$.art},228); continue;
        default: this.s=228; continue;
      }
      case 228: switch(this.lk()) {
        case 'returns': this.annotateReturns({outer:$.art},0); continue;
        case '{': this.annotateElementsBlock({art:$.art},229); continue;
        default: this.s=229; continue;
      }
      case 229: switch(this.lk()) {
        case 'actions': this.annotateActionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendArtifact($,$next) {
    this.rule_(231,$next);
    let name; let elemName; let elements; let incl; let _;
    for(;;) switch(this.s) {
      case 231: if(this.namePath(_={category:'Ext'},232)) name=_.name; continue;
      case 232: switch(this.lk()) {
        case ':': this.c(233); continue;
        case 'with': if(this.ck(242)) { this.addExtension( $.art, $.outer, 'extend', name );
this.docComment( $.art ); } continue;
        default: this.s=239;{ this.addExtension( $.art, $.outer, 'extend', name );
this.docComment( $.art ); } continue;
      }
      case 233: if(this.namePath(_={category:'ExtElement'},234)) {elemName=_.name; this.addExtension( $.art, $.outer, 'extend', name, elemName.path ); } continue;
      case 234: switch(this.lk()) {
        case 'with': this.ck(235); continue;
        default: this.s=235; continue;
      }
      case 235: this.s=236;{ this.docComment( $.art ); } continue;
      case 236: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},236); continue;
        default: this.s=237; continue;
      }
      case 237: switch(this.lk()) {
        case 'elements': if(this.ck(238)) elements=this.lb(); continue;
        case '{': this.s=238; continue;
        case 'enum': this.enumSymbolsBlock({art:$.art},0); continue;
        case '(': this.typeNamedArgsList({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 238: this.extendElementsBlock({art:$.art,start:elements},0); continue;
      case 239: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},239); continue;
        default: this.s=240; continue;
      }
      case 240: switch(this.l()) {
        case '{': this.extendElementsBlock({art:$.art,start:undefined},241); continue;
        default: this.gr([';']); continue;
      }
      case 241: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 242: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},242); continue;
        default: this.s=243; continue;
      }
      case 243: switch(this.lk()) {
        case 'Id': if(this.simplePath(_={category:'artref'},244)) {incl=_.ref; $.art.includes = [ incl ]; } continue;
        case 'elements': if(this.ckP(248,['{'])) elements=this.lb(); continue;
        case '{': this.s=248; continue;
        case 'actions': this.lP() && this.actionsBlock({art:$.art},0); continue;
        case 'enum': this.lP() && this.enumSymbolsBlock({art:$.art},0); continue;
        case '(': this.typeNamedArgsList({art:$.art},0); continue;
        case 'columns': this.ckP(250,['{']); continue;
        case 'definitions': this.ckP(251,['{']); continue;
        default: this.gi(0,[';']); continue;
      }
      case 244: switch(this.l()) {
        case ',': this.c(245); continue;
        default: this.s=246; continue;
      }
      case 245: if(this.simplePath(_={category:'artref'},244)) {incl=_.ref; $.art.includes.push( incl ); } continue;
      case 246: switch(this.l()) {
        case '{': this.extendElementsBlock({art:$.art,start:undefined},247); continue;
        default: this.s=247; continue;
      }
      case 247: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 248: this.extendElementsBlock({art:$.art,start:elements},249); continue;
      case 249: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 250: this.selectItemsList({query:$.art,start:this.lb()},0); continue;
      case 251: this.artifactsBlock({art:$.art,start:this.lb()},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendService($,$next) {
    this.rule_(253,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 253: if(this.mk(254,'service')) { $.art.expectedKind = this.valueWithLocation(); } continue;
      case 254: if(this.namePath(_={category:'ExtService'},255)) {name=_.name; $.art.name = name; $.outer.extensions.push( $.art ); } continue;
      case 255: switch(this.lk()) {
        case 'with': this.ck(256); continue;
        default: this.s=256; continue;
      }
      case 256: this.s=257;{ this.docComment( $.art ); } continue;
      case 257: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},257); continue;
        default: this.s=258; continue;
      }
      case 258: switch(this.l()) {
        case '{': this.artifactsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendContext($,$next) {
    this.rule_(260,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 260: if(this.mk(261,'context')) { $.art.expectedKind = this.valueWithLocation(); } continue;
      case 261: if(this.namePath(_={category:'ExtContext'},262)) {name=_.name; $.art.name = name; $.outer.extensions.push( $.art ); } continue;
      case 262: switch(this.lk()) {
        case 'with': this.ck(263); continue;
        default: this.s=263; continue;
      }
      case 263: this.s=264;{ this.docComment( $.art ); } continue;
      case 264: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},264); continue;
        default: this.s=265; continue;
      }
      case 265: switch(this.l()) {
        case '{': this.artifactsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendType($,$next) {
    this.rule_(267,$next);
    let name; let incl; let elements; let _;
    for(;;) switch(this.s) {
      case 267: if(this.mk(268,'type')) { $.art.expectedKind = this.valueWithLocation(); } continue;
      case 268: if(this.namePath(_={category:'Ext'},269)) {name=_.name; $.art.name = name; $.outer.extensions.push( $.art ); } continue;
      case 269: switch(this.lk()) {
        case 'with': if(this.ck(272)) { this.docComment( $.art ); } continue;
        default: this.s=270;{ this.docComment( $.art ); } continue;
      }
      case 270: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},270); continue;
        default: this.s=271; continue;
      }
      case 271: switch(this.l()) {
        case '{': this.extendElementsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      case 272: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},272); continue;
        default: this.s=273; continue;
      }
      case 273: switch(this.lk()) {
        case 'Id': if(this.simplePath(_={category:'artref'},274)) {incl=_.ref; $.art.includes = [ incl ]; } continue;
        case 'elements': if(this.ckP(277,['{'])) elements=this.lb(); continue;
        case '{': this.s=277; continue;
        case 'enum': this.lP() && this.enumSymbolsBlock({art:$.art},0); continue;
        case '(': this.typeNamedArgsList({art:$.art},0); continue;
        default: this.gi(0,[';']); continue;
      }
      case 274: switch(this.l()) {
        case ',': this.c(275); continue;
        default: this.s=276; continue;
      }
      case 275: if(this.simplePath(_={category:'artref'},274)) {incl=_.ref; $.art.includes.push( incl ); } continue;
      case 276: switch(this.l()) {
        case '{': this.extendElementsBlock({art:$.art,start:undefined},0); continue;
        default: this.gr([';']); continue;
      }
      case 277: this.extendElementsBlock({art:$.art,start:elements},0); continue;
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendEntityOrAspect($,$next) {
    this.rule_(279,$next);
    let name; let incl; let _;
    for(;;) switch(this.s) {
      case 279: switch(this.lk()) {
        case 'aspect': case 'entity': if(this.ck(280)) { $.art.expectedKind = this.valueWithLocation(); } continue;
        default: this.e(); continue;
      }
      case 280: if(this.namePath(_={category:'Ext'},281)) {name=_.name; $.art.name = name; $.outer.extensions.push( $.art ); } continue;
      case 281: switch(this.lk()) {
        case 'with': if(this.ck(283)) { this.docComment( $.art ); } continue;
        default: this.s=282;{ this.docComment( $.art ); } continue;
      }
      case 282: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},282); continue;
        default: this.s=287; continue;
      }
      case 283: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},283); continue;
        default: this.s=284; continue;
      }
      case 284: switch(this.lk()) {
        case 'Id': if(this.simplePath(_={category:'artref'},285)) {incl=_.ref; $.art.includes = [ incl ]; } continue;
        case 'actions': this.gP(287); continue;
        default: this.gi(287); continue;
      }
      case 285: switch(this.l()) {
        case ',': this.c(286); continue;
        default: this.s=287; continue;
      }
      case 286: if(this.simplePath(_={category:'artref'},285)) {incl=_.ref; $.art.includes.push( incl ); } continue;
      case 287: switch(this.l()) {
        case '{': this.extendElementsBlock({art:$.art,start:undefined},288); continue;
        default: this.s=288; continue;
      }
      case 288: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  extendProjection($,$next) {
    this.rule_(290,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 290: if(this.mk(291,'projection')) { $.art.expectedKind = this.valueWithLocation(); } continue;
      case 291: if(this.namePath(_={category:'Ext'},292)) {name=_.name; $.art.name = name; $.outer.extensions.push( $.art ); } continue;
      case 292: switch(this.lk()) {
        case 'with': this.ck(293); continue;
        default: this.s=293; continue;
      }
      case 293: this.s=294;{ this.docComment( $.art ); } continue;
      case 294: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},294); continue;
        default: this.s=295; continue;
      }
      case 295: switch(this.l()) {
        case '{': this.selectItemsList({query:$.art,start:undefined},296); continue;
        default: this.s=296; continue;
      }
      case 296: switch(this.lk()) {
        case 'actions': this.actionsBlock({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( $.art );
      return this.exit_();
    }
  }
  annotateActionsBlock($,$next) {
    this.rule_(298,$next);
    for(;;) switch(this.s) {
      case 298: if(this.mk(299,'actions')) { $.art.actions = this.createDict(); } continue;
      case 299: this.m(300,'{'); continue;
      case 300: switch(this.l()) {
        case 'Id': case '@': this.annotateBoundAction({outer:$.art},301); continue;
        default: this.s=302; continue;
      }
      case 301: switch(this.l()) {
        case ';': this.c(300); continue;
        case '}': this.s=302; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(300); continue;
        default: this.e(); continue;
      }
      case 302: if(this.m(0,'}')) {this.afterBrace(); this.finalizeExtensionsDict( $.art.actions ); } continue;
      default: return this.exit_();
    }
  }
  annotateBoundAction($,$next) {
    this.rule_(304,$next);
    let art=new XsnArtifact();
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 304: switch(this.l()) {
        case '@': this.annoAssignStd({art},304); continue;
        default: this.s=305; continue;
      }
      case 305: if(this.mi(306,'ExtBoundAction')) { this.addDef( art, $.outer, 'actions', 'annotate', this.identAst() );
this.docComment( art ); } continue;
      case 306: switch(this.l()) {
        case '@': this.annoAssignStd({art},306); continue;
        default: this.s=307; continue;
      }
      case 307: switch(this.l()) {
        case '(': this.annotateParamsBlock({art},308); continue;
        default: this.s=308; continue;
      }
      case 308: switch(this.lk()) {
        case 'returns': this.lP() && this.annotateReturns({outer:art},0); continue;
        default: this.gr(['Id',';','@','}']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  annotateParamsBlock($,$next) {
    this.rule_(310,$next);
    for(;;) switch(this.s) {
      case 310: if(this.m(311,'(')) { $.art.params = this.createDict(); } continue;
      case 311: switch(this.l()) {
        case 'Id': case '@': this.annotateParam({outer:$.art},312); continue;
        default: this.s=313; continue;
      }
      case 312: switch(this.l()) {
        case ',': this.c(311); continue;
        case ')': this.s=313; continue;
        default: this.e(); continue;
      }
      case 313: if(this.m(0,')')) { this.finalizeExtensionsDict( $.art.params ); } continue;
      default: return this.exit_();
    }
  }
  annotateParam($,$next) {
    this.rule_(315,$next);
    let art=new XsnArtifact();
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 315: switch(this.l()) {
        case '@': this.annoAssignStd({art},315); continue;
        default: this.s=316; continue;
      }
      case 316: if(this.mi(317,'ExtParam')) { this.addDef( art, $.outer, 'params', 'annotate', this.identAst() );
this.docComment( art ); } continue;
      case 317: switch(this.l()) {
        case '@': this.annoAssignStd({art},317); continue;
        default: this.gr([')',',']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  annotateReturns($,$next) {
    this.rule_(319,$next);
    let art=new XsnArtifact();
    for(;;) switch(this.s) {
      case 319: if(this.mk(320,'returns')) { $.outer.returns = art; art.kind = 'annotate';
this.docComment( art ); } continue;
      case 320: switch(this.l()) {
        case '@': this.annoAssignStd({art},320); continue;
        default: this.s=321; continue;
      }
      case 321: switch(this.l()) {
        case '{': this.annotateElementsBlock({art},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  annotateElementsBlock($,$next) {
    this.rule_(323,$next);
    for(;;) switch(this.s) {
      case 323: if(this.m(324,'{')) { $.art.elements = this.createDict(); } continue;
      case 324: switch(this.l()) {
        case 'Id': case '@': this.annotateElement({outer:$.art},325); continue;
        default: this.s=326; continue;
      }
      case 325: switch(this.l()) {
        case ';': this.c(324); continue;
        case '}': this.s=326; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(324); continue;
        default: this.e(); continue;
      }
      case 326: if(this.m(0,'}')) {this.afterBrace(); this.finalizeExtensionsDict( $.art.elements ); } continue;
      default: return this.exit_();
    }
  }
  annotateElement($,$next) {
    this.rule_(328,$next);
    let art=new XsnArtifact();
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 328: switch(this.l()) {
        case '@': this.annoAssignStd({art},328); continue;
        default: this.s=329; continue;
      }
      case 329: if(this.mi(330,'ExtElement')) { this.addDef( art, $.outer, 'elements', 'annotate', this.identAst() );
this.docComment( art ); } continue;
      case 330: switch(this.l()) {
        case '@': this.annoAssignStd({art},330); continue;
        default: this.s=331; continue;
      }
      case 331: switch(this.l()) {
        case '{': this.annotateElementsBlock({art},0); continue;
        default: this.gr(['Id',';','@','}']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  extendElementsBlock($,$next) {
    this.rule_(333,$next);
    for(;;) switch(this.s) {
      case 333: if(this.m(334,'{')) { $.art.elements = this.createDict( $.start ); } continue;
      case 334: switch(this.l()) {
        case 'Id': case '@': this.elementDefOrExtend({outer:$.art},335); continue;
        default: this.s=336; continue;
      }
      case 335: switch(this.l()) {
        case ';': this.c(334); continue;
        case '}': this.s=336; continue;
        case 'Id': case '@': this.ec('afterBrace') && this.g(334); continue;
        default: this.e(); continue;
      }
      case 336: if(this.m(0,'}')) {this.afterBrace(); this.finalizeExtensionsDict( $.art.elements ); } continue;
      default: return this.exit_();
    }
  }
  elementDefOrExtend($,$next) {
    this.rule_(338,$next);
    let art=new XsnArtifact();
    let elements;
    { this.docComment( art ); }
    for(;;) switch(this.s) {
      case 338: switch(this.l()) {
        case '@': this.annoAssignStd({art},338); continue;
        default: this.s=339; continue;
      }
      case 339: switch(this.lk()) {
        case 'Id': case '@': case 'key': this.elementDef({outer:$.outer,art},0); continue;
        case 'extend': this.ckP(340,['Id']); continue;
        default: this.ei(); continue;
      }
      case 340: switch(this.lk()) {
        case 'element': if(this.ckP(341,['Id'])) { art.expectedKind = this.valueWithLocation(); } continue;
        default: this.s=341; continue;
      }
      case 341: if(this.mi(342,'ExtElement')) { this.addDef( art, $.outer, 'elements', 'extend', this.identAst() ); } continue;
      case 342: switch(this.lk()) {
        case 'with': if(this.ck(345)) { this.docComment( art ); } continue;
        default: this.s=343;{ this.docComment( art ); } continue;
      }
      case 343: switch(this.l()) {
        case '@': this.annoAssignStd({art},343); continue;
        default: this.s=344; continue;
      }
      case 344: switch(this.l()) {
        case '{': this.extendElementsBlock({art,start:undefined},0); continue;
        default: this.gr(['Id',';','@','}','key']); continue;
      }
      case 345: switch(this.l()) {
        case '@': this.annoAssignStd({art},345); continue;
        default: this.s=346; continue;
      }
      case 346: switch(this.lk()) {
        case 'elements': if(this.ck(347)) elements=this.lb(); continue;
        case '{': this.s=347; continue;
        case 'enum': this.enumSymbolsBlock({art},0); continue;
        case '(': this.typeNamedArgsList({art},0); continue;
        default: this.gr(['Id',';','@','}','key']); continue;
      }
      case 347: this.extendElementsBlock({art,start:elements},0); continue;
      default: return this.exit_();
    }
  }
  typeOrIncludesSpec($,$next) {
    this.rule_(349,$next);
    let ref; let _;
    for(;;) switch(this.s) {
      case 349: switch(this.l()) {
        case '{': this.elementsBlock({art:$.art},350); continue;
        case ':': this.c(351); continue;
        default: this.e(); continue;
      }
      case 350: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      case 351: switch(this.lk()) {
        case '{': case 'many': case 'type': case 'array': case 'localized': case 'association': case 'composition': this.lP() && this.typeExpression({art:$.art},0); continue;
        case 'Id': if(this.simplePath(_={category:'artref'},352)) {ref=_.ref; $.art.type = ref; } continue;
        default: this.ei(); continue;
      }
      case 352: switch(this.lk()) {
        case '(': case ':': this.typeRefOptArgs({art:$.art},353,410); continue;
        case '@': case 'not': case 'enum': case 'null': case 'default': this.typeExpression({art:$.art},0,362); continue;
        case ',': case '{': this.s=354;{ $.art.includes = [ $.art.type ]; delete $.art.type; } continue;
        default: if(this.gr([';'])) { this.docComment( $.art ); } continue;
      }
      case 353: switch(this.lk()) {
        case '@': case 'not': case 'enum': case 'null': case 'default': this.typeExpression({art:$.art},0,362); continue;
        default: if(this.gr([';'])) { this.docComment( $.art ); } continue;
      }
      case 354: switch(this.l()) {
        case ',': this.c(355); continue;
        default: this.s=357; continue;
      }
      case 355: switch(this.l()) {
        case 'Id': if(this.simplePath(_={category:'artref'},356)) {ref=_.ref; $.art.includes.push( ref ); } continue;
        default: this.s=357; continue;
      }
      case 356: switch(this.l()) {
        case ',': this.c(355); continue;
        case '{': this.s=357; continue;
        default: this.e(); continue;
      }
      case 357: this.elementsBlock({art:$.art},358); continue;
      case 358: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},0); continue;
        default: this.gr([';']); continue;
      }
      default: return this.exit_();
    }
  }
  typeExpression($,$next,$startState) {
    this.rule_($startState??360,$next);
    let assoc; let card; let target; let cond; let _;
    for(;;) switch(this.s) {
      case 360: switch(this.lk()) {
        case '{': this.elementsBlock({art:$.art},361); continue;
        case 'Id': this.typeRefOptArgs({art:$.art},362); continue;
        case 'type': this.lP() && this.typeTypeOf({art:$.art},362); continue;
        case 'localized': if(this.ckP(370,['Id'])) { $.art.localized = this.valueWithLocation( true ); } continue;
        case 'association': if(this.ckP(374,['[','to'])) assoc=this.lb(); continue;
        case 'composition': if(this.ckP(383,['[','of'])) assoc=this.lb(); continue;
        case 'array': this.ckP(393,['of']); continue;
        case 'many': if(this.ckP(394,['Id','{'])) { $.art.items = { location: this.lb().location }; } continue;
        default: this.ei(); continue;
      }
      case 361: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},0); continue;
        default: this.gr([]); continue;
      }
      case 362: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},363); continue;
        default: this.s=363; continue;
      }
      case 363: this.s=364;{ this.calcOrDefaultRestriction(); } continue;
      case 364: switch(this.lk()) {
        case 'not': case 'null': case 'default': if(this.nullabilityAndDefault({art:$.art},365)) { this.docComment( $.art ); } continue;
        case 'enum': this.enumSymbolsBlock({art:$.art},366); continue;
        case '@': this.s=367;{ this.docComment( $.art ); } continue;
        default: if(this.gr([])) { this.docComment( $.art ); } continue;
      }
      case 365: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},365); continue;
        default: this.gr([]); continue;
      }
      case 366: switch(this.lk()) {
        case 'not': case 'null': case 'default': this.nullabilityAndDefault({art:$.art},0); continue;
        default: this.gr([]); continue;
      }
      case 367: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},367); continue;
        default: this.s=368; continue;
      }
      case 368: switch(this.lk()) {
        case 'enum': if(this.enumSymbolsBlock({art:$.art},369)) { this.calcOrDefaultRestriction(); } continue;
        default: this.gr([]); continue;
      }
      case 369: switch(this.lk()) {
        case 'not': case 'null': case 'default': this.nullabilityAndDefault({art:$.art},0); continue;
        default: this.gr([]); continue;
      }
      case 370: if(this.typeRefOptArgs({art:$.art},371)) { this.calcOrDefaultRestriction(); } continue;
      case 371: switch(this.lk()) {
        case 'not': case 'null': case 'default': this.nullabilityAndDefault({art:$.art},372); continue;
        default: this.s=372; continue;
      }
      case 372: this.s=373;{ this.docComment( $.art ); } continue;
      case 373: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},373); continue;
        default: this.gr([]); continue;
      }
      case 374: switch(this.l()) {
        case '[': this.cardinality({art:$.art},375); continue;
        default: this.s=375; continue;
      }
      case 375: this.mk(376,'to'); continue;
      case 376: switch(this.lk()) {
        case 'one': case 'many': if(this.ckP(377,['Id'])) card=this.lb(); continue;
        default: this.s=377; continue;
      }
      case 377: if(this.simplePath(_={category:'artref'},378)) {target=_.ref; this.setAssocAndComposition( $.art, assoc, card, target );
this.calcOrDefaultRestriction(false,true); } continue;
      case 378: switch(this.lk()) {
        case 'on': this.ck(379); continue;
        case '{': this.foreignKeysBlock({art:$.art},380); continue;
        default: this.s=380; continue;
      }
      case 379: if(this.condition(_={},381)) {cond=_.expr; $.art.on = cond; } continue;
      case 380: switch(this.lk()) {
        case 'not': case 'null': case 'default': this.nullabilityAndDefault({art:$.art},381); continue;
        default: this.s=381; continue;
      }
      case 381: this.s=382;{ this.docComment( $.art ); } continue;
      case 382: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},382); continue;
        default: this.gr([]); continue;
      }
      case 383: switch(this.l()) {
        case '[': this.cardinality({art:$.art},384); continue;
        default: this.s=384; continue;
      }
      case 384: this.mk(385,'of'); continue;
      case 385: switch(this.lk()) {
        case 'one': case 'many': if(this.ckP(386,['Id','{'])) card=this.lb(); continue;
        default: this.s=386; continue;
      }
      case 386: switch(this.l()) {
        case 'Id': if(this.simplePath(_={category:'artref'},387)) {target=_.ref; this.setAssocAndComposition( $.art, assoc, card, target );
this.calcOrDefaultRestriction(false,true); } continue;
        case '{': this.s=392;{ target = {}; this.setAssocAndComposition( $.art, assoc, card, target ); } continue;
        default: this.e(); continue;
      }
      case 387: switch(this.lk()) {
        case 'on': this.ck(388); continue;
        case '{': this.foreignKeysBlock({art:$.art},389); continue;
        default: this.s=389; continue;
      }
      case 388: if(this.condition(_={},390)) {cond=_.expr; $.art.on = cond; } continue;
      case 389: switch(this.lk()) {
        case 'not': case 'null': case 'default': this.nullabilityAndDefault({art:$.art},390); continue;
        default: this.s=390; continue;
      }
      case 390: this.s=391;{ this.docComment( $.art ); } continue;
      case 391: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},391); continue;
        default: this.gr([]); continue;
      }
      case 392: if(this.elementsBlock({art:target},0)) { target.location = target.elements[Symbol.for('cds.$location')]; } continue;
      case 393: if(this.mk(394,'of')) { $.art.items = { location: this.locationOfPrevTokens( 2 ) }; } continue;
      case 394: switch(this.lk()) {
        case '{': this.elementsBlock({art:$.art.items},395); continue;
        case 'Id': this.typeRefOptArgs({art:$.art.items},396); continue;
        case 'type': this.lP() && this.typeTypeOf({art:$.art.items},396); continue;
        default: this.ei(); continue;
      }
      case 395: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art.items},0); continue;
        default: this.gr([]); continue;
      }
      case 396: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art.items},397); continue;
        default: this.s=397; continue;
      }
      case 397: switch(this.lk()) {
        case 'enum': this.enumSymbolsBlock({art:$.art.items},398); continue;
        default: this.s=399;{ this.docComment( $.art ); } continue;
      }
      case 398: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art.items},0); continue;
        default: this.gr([]); continue;
      }
      case 399: switch(this.l()) {
        case '@': this.annoAssignStd({art:$.art},399); continue;
        default: this.gr([]); continue;
      }
      default: return this.exit_();
    }
  }
  typeTypeOf($,$next) {
    this.rule_(401,$next);
    let location;
    let type; let _;
    for(;;) switch(this.s) {
      case 401: this.mk(402,'type'); continue;
      case 402: if(this.mk(403,'of')) { location = this.locationOfPrevTokens( 2 ); } continue;
      case 403: if(this.simplePath(_={category:'ref'},404)) {type=_.ref; $.art.type = type; } continue;
      case 404: switch(this.l()) {
        case ':': if(this.c(405)) { type.scope = type.path.length; } continue;
        default: if(this.gr(['@'])) { type.scope = 'typeOf'; type.path.unshift( { id: 'type of', location } ); } continue;
      }
      case 405: if(this.miA(406,'ref')) { type.path.push( this.identAst() ); } continue;
      case 406: switch(this.l()) {
        case '.': this.c(407); continue;
        default: this.gr(['@']); continue;
      }
      case 407: if(this.miA(406,'ref')) { type.path.push( this.identAst() ); } continue;
      default: return this.exit_();
    }
  }
  typeRefOptArgs($,$next,$startState) {
    this.rule_($startState??409,$next);
    let type=$.art.type;
    let open; let tok; let _;
    for(;;) switch(this.s) {
      case 409: if(this.simplePath(_={category:'artref'},410)) {type=_.ref; $.art.type = type; } continue;
      case 410: switch(this.l()) {
        case ':': if(this.c(411)) { type.scope = type.path.length; } continue;
        case '(': if(this.c(414)) open=this.lb(); continue;
        default: this.gr([]); continue;
      }
      case 411: if(this.miA(412,'ref')) { type.path.push( this.identAst() ); } continue;
      case 412: switch(this.l()) {
        case '.': this.c(413); continue;
        default: this.gr([]); continue;
      }
      case 413: if(this.miA(412,'ref')) { type.path.push( this.identAst() ); } continue;
      case 414: switch(this.lk()) {
        case 'Number': if(this.c(415)) { $.art.$typeArgs = this.createArray( open );
$.art.$typeArgs.push( this.unsignedIntegerLiteral() ); } continue;
        case 'Id': this.s=417;{ $.art.$typeArgs = this.createDict( open ); } continue;
        default: this.ei(); continue;
      }
      case 415: switch(this.l()) {
        case ',': this.c(416); continue;
        default: this.s=419; continue;
      }
      case 416: switch(this.lk()) {
        case 'Number': if(this.c(415)) { $.art.$typeArgs.push( this.unsignedIntegerLiteral() ); } continue;
        case 'floating': case 'variable': if(this.ck(415)) {tok=this.lb(); $.art.$typeArgs.push( { literal: 'string', val: tok.keyword, location: tok.location } ); } continue;
        case ')': this.s=419; continue;
        default: this.e(); continue;
      }
      case 417: switch(this.l()) {
        case 'Id': this.typeNamedArg({art:$.art},418); continue;
        default: this.s=419; continue;
      }
      case 418: switch(this.l()) {
        case ',': this.c(417); continue;
        case ')': this.s=419; continue;
        default: this.e(); continue;
      }
      case 419: if(this.m(0,')')) { this.finalizeDictOrArray( $.art.$typeArgs ); } continue;
      default: return this.exit_();
    }
  }
  typeNamedArgsList($,$next) {
    this.rule_(421,$next);
    for(;;) switch(this.s) {
      case 421: if(this.m(422,'(')) { $.art.$typeArgs = this.createDict(); } continue;
      case 422: switch(this.l()) {
        case 'Id': this.typeNamedArg({art:$.art},423); continue;
        default: this.s=424; continue;
      }
      case 423: switch(this.l()) {
        case ',': this.c(422); continue;
        case ')': this.s=424; continue;
        default: this.e(); continue;
      }
      case 424: if(this.m(0,')')) { this.finalizeDictOrArray( $.art.$typeArgs ); } continue;
      default: return this.exit_();
    }
  }
  typeNamedArg($,$next) {
    this.rule_(426,$next);
    let name; let tok;
    for(;;) switch(this.s) {
      case 426: if(this.mi(427,'typeparamname')) name=this.lb(); continue;
      case 427: this.m(428,':'); continue;
      case 428: switch(this.lk()) {
        case 'Number': if(this.c(0)) { this.setTypeFacet( $.art, name, this.unsignedIntegerLiteral() ); } continue;
        case 'floating': case 'variable': if(this.ck(0)) {tok=this.lb(); this.setTypeFacet( $.art, name, { literal: 'string', val: tok.keyword, location: tok.location } ); } continue;
        default: this.e(); continue;
      }
      default: return this.exit_();
    }
  }
  cardinality($,$next) {
    this.rule_(430,$next);
    let card={};
    for(;;) switch(this.s) {
      case 430: this.m(431,'['); continue;
      case 431: switch(this.l()) {
        case '*': if(this.c(432)) { card.targetMax = this.valueWithLocation(); } continue;
        case 'Number': if(this.c(434)) { card.targetMax = this.unsignedIntegerLiteral(); } continue;
        default: this.s=436;{ card.targetMax = this.valueWithLocation( '*' ); } continue;
      }
      case 432: switch(this.l()) {
        case ',': this.c(433); continue;
        default: this.s=436; continue;
      }
      case 433: this.targetCardinality({card,atAlt:false},436); continue;
      case 434: switch(this.l()) {
        case ',': this.c(435); continue;
        case '..': this.targetCardinality({card,atAlt:true},436,439); continue;
        default: this.s=436; continue;
      }
      case 435: this.targetCardinality({card,atAlt:false},436); continue;
      case 436: this.m(0,']'); continue;
      default:
      $.art.cardinality = this.attachLocation(card);
      return this.exit_();
    }
  }
  targetCardinality($,$next,$startState) {
    this.rule_($startState??438,$next);
    { if (!$.atAlt) $.card.sourceMax = $.card.targetMax; }
    for(;;) switch(this.s) {
      case 438: switch(this.l()) {
        case '*': if(this.c(0)) { $.card.targetMax = this.valueWithLocation(); } continue;
        case 'Number': if(this.c(439)) { $.card.targetMax = this.unsignedIntegerLiteral(); } continue;
        default: this.e(); continue;
      }
      case 439: switch(this.l()) {
        case '..': if(this.c(440)) { $.card.targetMin = $.card.targetMax; } continue;
        default: this.gr([']']); continue;
      }
      case 440: switch(this.l()) {
        case '*': if(this.c(0)) { $.card.targetMax = this.valueWithLocation(); } continue;
        case 'Number': if(this.c(0)) { $.card.targetMax = this.unsignedIntegerLiteral(); } continue;
        default: this.e(); continue;
      }
      default: return this.exit_();
    }
  }
  nullabilityAndDefault($,$next) {
    this.rule_(442,$next);
    let expr; let _;
    for(;;) switch(this.s) {
      case 442: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},443); continue;
        case 'default': this.ec('calcOrDefaultRestriction') && this.ck(445); continue;
        default: this.e(); continue;
      }
      case 443: switch(this.lk()) {
        case 'default': this.gc(0,'calcOrDefaultRestriction') && this.ck(444); continue;
        default: this.gr([]); continue;
      }
      case 444: if(this.expression(_={},0)) {expr=_.expr; $.art.default = expr; } continue;
      case 445: if(this.expression(_={},446)) {expr=_.expr; $.art.default = expr; } continue;
      case 446: switch(this.lk()) {
        case 'not': case 'null': this.nullability({art:$.art},0); continue;
        default: this.gr([]); continue;
      }
      default: return this.exit_();
    }
  }
  nullability($,$next) {
    this.rule_(448,$next);
    for(;;) switch(this.s) {
      case 448: switch(this.lk()) {
        case 'null': if(this.ck(0)) { this.setNullability( $.art, false ); } continue;
        case 'not': this.ck(449); continue;
        default: this.e(); continue;
      }
      case 449: if(this.mk(0,'null')) { this.setNullability( $.art, true, this.locationOfPrevTokens( 2 ) ); } continue;
      default: return this.exit_();
    }
  }
  queryEOF($,$next) {
    $.query??=undefined;
    this.rule_(451,$next);
    let _;
    for(;;) switch(this.s) {
      case 451: if(this.queryExpression(_={},452)) $.query=_.expr; continue;
      case 452: switch(this.l()) {
        case ';': this.c(453); continue;
        default: this.s=453; continue;
      }
      case 453: this.m(0,'EOF'); continue;
      default: return this.exit_();
    }
  }
  projectionSpec($,$next) {
    $.query??={};
    this.rule_(455,$next);
    let tab; let _;
    for(;;) switch(this.s) {
      case 455: if(this.mk(456,'projection')) { $.query = { op: this.valueWithLocation( 'SELECT' ) }; } continue;
      case 456: this.mk(457,'on'); continue;
      case 457: if(this.fromRefWithOptAlias(_={},458)) {tab=_.expr; this.afterBrace();
$.query.from = tab; } continue;
      case 458: switch(this.l()) {
        case '{': this.selectItemsList({query:$.query,start:undefined},459); continue;
        default: this.s=459; continue;
      }
      case 459: switch(this.lk()) {
        case 'excluding': this.excludingClause({query:$.query},0); continue;
        default: this.gr([';']); continue;
      }
      default:
      this.attachLocation($.query);
      return this.exit_();
    }
  }
  queryExpression($,$next,$startState) {
    $.expr??={};
    this.rule_($startState??461,$next);
    let op; let quantifier;
    let query; let _;
    for(;;) switch(this.s) {
      case 461: switch(this.l()) {
        case '(': this.c(462); continue;
        case 'Id': if(this.selectQuery(_={},464)) $.expr=_.query; continue;
        default: this.e(); continue;
      }
      case 462: this.queryExpression($,463); continue;
      case 463: if(this.m(464,')')) { this.surroundByParens( $.expr ); } continue;
      case 464: switch(this.lk()) {
        case 'intersect': if(this.gp(468,4) && this.ck(465)) { op = this.valueWithLocation(); } continue;
        case 'minus': case 'except': if(this.gp(468,2) && this.ck(465)) { op = this.valueWithLocation(); } continue;
        case 'union': if(this.gp(468,2) && this.ck(466)) { op = this.valueWithLocation(); } continue;
        default: this.s=468; continue;
      }
      case 465: switch(this.lk()) {
        case 'distinct': if(this.ck(467)) { quantifier = this.valueWithLocation(); } continue;
        default: this.s=467; continue;
      }
      case 466: switch(this.lk()) {
        case 'all': case 'distinct': if(this.ck(467)) { quantifier = this.valueWithLocation(); } continue;
        default: this.s=467; continue;
      }
      case 467: if(this.queryExpression(_={},464)) {query=_.expr; if ($.expr.$parens || op.val !== $.expr.op.val || quantifier?.val !== $.expr.quantifier?.val) $.expr = { op, args: [$.expr], quantifier };
quantifier = undefined;
$.expr.args.push( query ); this.attachLocation( $.expr ); } continue;
      case 468: switch(this.lk()) {
        case 'limit': case 'order': if(this.gp(0,0,'post') && this.g(469)) { if ($.expr.$parens) { this.attachLocation( $.expr ); $.expr = { op: this.valueWithLocation( '$query', this.la() ), args: [ $.expr ] }; } } continue;
        default: this.gr([]); continue;
      }
      case 469: this.orderByLimitOffset({query:$.expr},0); continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  selectQuery($,$next) {
    $.query??={};
    this.rule_(471,$next);
    for(;;) switch(this.s) {
      case 471: if(this.mk(472,'select')) { $.query = { op: this.valueWithLocation( 'SELECT' ) }; } continue;
      case 472: switch(this.lk()) {
        case 'from': this.ck(473); continue;
        case 'all': case 'distinct': if(this.ckP(483,['Id','#','(','*','+','-',':','?','@','{','key','not','case','cast','null','true','false','Number','String','exists','QuotedLiteral'])) { $.query.quantifier = this.valueWithLocation(); } continue;
        case 'Id': case '#': case '(': case '*': case '+': case '-': case ':': case '?': case '@': case '{': case 'key': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': this.s=483; continue;
        default: this.ei(); continue;
      }
      case 473: this.querySource({query:$.query},474); continue;
      case 474: switch(this.lk()) {
        case 'mixin': this.ck(475); continue;
        default: this.s=480; continue;
      }
      case 475: if(this.m(476,'{')) { $.query.mixin = this.createDict(); } continue;
      case 476: switch(this.l()) {
        case 'Id': this.mixinElementDef({outer:$.query},477); continue;
        default: this.s=478; continue;
      }
      case 477: switch(this.l()) {
        case ';': this.c(476); continue;
        case '}': this.s=478; continue;
        default: this.e(); continue;
      }
      case 478: if(this.m(479,'}')) { this.finalizeDictOrArray( $.query.mixin ); } continue;
      case 479: this.mk(480,'into'); continue;
      case 480: switch(this.lk()) {
        case 'all': case 'distinct': if(this.ck(481)) { $.query.quantifier = this.valueWithLocation(); } continue;
        default: this.s=481; continue;
      }
      case 481: switch(this.l()) {
        case '{': this.selectItemsList({query:$.query,start:undefined},482); continue;
        default: this.s=482; continue;
      }
      case 482: switch(this.lk()) {
        case 'excluding': this.excludingClause({query:$.query},488); continue;
        default: this.s=488; continue;
      }
      case 483: switch(this.l()) {
        case '*': if(this.c(484)) { $.query.columns = [ this.valueWithLocation() ]; } continue;
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case '@': case '{': case 'Number': case 'String': case 'QuotedLiteral': this.selectItemDef({columns:($.query.columns = [])},484); continue;
        default: this.e(); continue;
      }
      case 484: switch(this.l()) {
        case ',': this.c(485); continue;
        default: this.s=486; continue;
      }
      case 485: switch(this.l()) {
        case '*': if(this.c(484)) { $.query.columns.push( this.valueWithLocation() ); } continue;
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case '@': case '{': case 'Number': case 'String': case 'QuotedLiteral': this.selectItemDef({columns:$.query.columns},484); continue;
        default: this.e(); continue;
      }
      case 486: this.mk(487,'from'); continue;
      case 487: this.querySource({query:$.query},488); continue;
      case 488: switch(this.lk()) {
        case 'group': case 'where': case 'having': this.whereGroupByHaving({query:$.query},0); continue;
        default: this.gr(['limit','minus','order','union','except','intersect']); continue;
      }
      default:
      this.attachLocation($.query);
      return this.exit_();
    }
  }
  querySource($,$next) {
    this.rule_(490,$next);
    let tab; let _;
    for(;;) switch(this.s) {
      case 490: if(this.tableExpression(_={},491)) {tab=_.expr; $.query.from = tab; } continue;
      case 491: switch(this.l()) {
        case ',': this.s=492;{ const { location } = this.la();
$.query.from = { op: { val: 'join', location }, join: { val: 'cross', location }, args: [tab] }; } continue;
        default: this.gr(['group','limit','minus','order','union','where','except','having','intersect']); continue;
      }
      case 492: switch(this.l()) {
        case ',': this.c(493); continue;
        default: this.gr(['group','limit','minus','order','union','where','except','having','intersect']); continue;
      }
      case 493: if(this.tableExpression(_={},492)) {tab=_.expr; $.query.from.args.push( tab ); } continue;
      case 0:
      this.attachLocation($.query.from);
      default: return this.exit_();
    }
  }
  tableExpression($,$next,$startState) {
    $.expr??=undefined;
    this.rule_($startState??495,$next);
    let join; let tab; let card; let cond; let _;
    for(;;) switch(this.s) {
      case 495: switch(this.l()) {
        case '(': this.tableOrQueryParens($,496); continue;
        case 'Id': this.fromRefWithOptAlias($,497); continue;
        default: this.e(); continue;
      }
      case 496: this.s=497;{ $.expr = this.taggedIfQuery( $.expr ); } continue;
      case 497: switch(this.lk()) {
        case 'cross': if(this.ck(498)) join=this.lb(); continue;
        case 'inner': if(this.ck(501)) join=this.lb(); continue;
        case 'full': case 'left': case 'right': if(this.ck(500)) join=this.lb(); continue;
        case 'join': this.s=502; continue;
        default: if(this.gr([])) { join = undefined; } continue;
      }
      case 498: if(this.mk(499,'join')) { if ($.expr?.join?.val !== 'cross' || $.expr.$parens) $.expr = { op: this.valueWithLocation(), join: this.valueWithLocation( undefined, join ), args: [ $.expr ] }; } continue;
      case 499: switch(this.l()) {
        case '(': if(this.tableOrQueryParens(_={},497)) {tab=_.expr; $.expr.args.push( this.taggedIfQuery( tab ) );
this.attachLocation( $.expr ); } continue;
        case 'Id': if(this.fromRefWithOptAlias(_={},497)) {tab=_.expr; $.expr.args.push( tab );
this.attachLocation( $.expr ); } continue;
        default: this.e(); continue;
      }
      case 500: switch(this.lk()) {
        case 'outer': this.ck(501); continue;
        default: this.s=501; continue;
      }
      case 501: switch(this.lk()) {
        case 'one': case 'many': case 'exact': this.joinCardinality(card={},502); continue;
        default: this.s=502; continue;
      }
      case 502: if(this.mk(503,'join')) { $.expr = { op: this.valueWithLocation(), join: this.valueWithLocation( join?.keyword || 'inner', join ), args: [ $.expr ] }; if (card) $.expr.cardinality = card; card = undefined;
join = undefined; } continue;
      case 503: if(this.tableExpression(_={},504)) {tab=_.expr; $.expr.args.push( tab ); this.attachLocation( $.expr ); } continue;
      case 504: this.mk(505,'on'); continue;
      case 505: if(this.condition(_={},497)) {cond=_.expr; $.expr.on = cond;
this.attachLocation( $.expr ); } continue;
      default: return this.exit_();
    }
  }
  tableOrQueryParens($,$next) {
    $.expr??=undefined;
    this.rule_(507,$next);
    for(;;) switch(this.s) {
      case 507: this.m(508,'('); continue;
      case 508: switch(this.lk()) {
        case '(': this.tableOrQueryParens($,509); continue;
        case 'Id': this.ep(-2) && this.tableExpression($,510); continue;
        case 'select': this.ep(-1) && this.queryExpression($,510); continue;
        default: this.ei(); continue;
      }
      case 509: switch(this.lk()) {
        case 'full': case 'join': case 'left': case 'cross': case 'inner': case 'right': this.gp(510,-2,'post') && this.tableExpression($,510,496); continue;
        case 'limit': case 'minus': case 'order': case 'union': case 'except': case 'intersect': this.gp(510,-1,'post') && this.queryExpression($,510,464); continue;
        default: this.s=510; continue;
      }
      case 510: if(this.m(511,')')) { this.surroundByParens( $.expr ); } continue;
      case 511: switch(this.lk()) {
        case 'as': this.gp(0,-2,'none') && this.ck(512); continue;
        case 'Id': if(this.gc(0,'tableAlias') && this.ci(0,'FromAlias')) {this.setPrecInCallingRule(); $.expr = this.taggedIfQuery( $.expr ); $.expr.name = this.fragileAlias(); } continue;
        default: if(this.gi(0,['full','join','left','cross','inner','right'])) { this.setPrecInCallingRule(); } continue;
      }
      case 512: if(this.mi(0,'FromAlias')) {this.setPrecInCallingRule(); $.expr = this.taggedIfQuery( $.expr ); $.expr.name = this.identAst(); } continue;
      default: return this.exit_();
    }
  }
  joinCardinality($,$next) {
    $.sourceMax??=undefined; $.targetMax??=undefined;
    this.rule_(514,$next);
    for(;;) switch(this.s) {
      case 514: switch(this.lk()) {
        case 'exact': if(this.ck(515)) { $.sourceMin = this.valueWithLocation( 1 ); } continue;
        case 'one': this.s=515; continue;
        case 'many': if(this.ck(516)) { $.sourceMax = this.valueWithLocation( '*' ); } continue;
        default: this.e(); continue;
      }
      case 515: if(this.mk(516,'one')) { $.sourceMax = this.valueWithLocation( 1 ); } continue;
      case 516: this.mk(517,'to'); continue;
      case 517: switch(this.lk()) {
        case 'exact': if(this.ck(518)) { $.targetMin = this.valueWithLocation( 1 ); } continue;
        case 'one': this.s=518; continue;
        case 'many': if(this.ck(0)) { $.targetMax = this.valueWithLocation( '*' ); } continue;
        default: this.e(); continue;
      }
      case 518: if(this.mk(0,'one')) { $.targetMax = this.valueWithLocation( 1 ); } continue;
      default:
      this.attachLocation( $ );
      return this.exit_();
    }
  }
  fromRefWithOptAlias($,$next) {
    $.expr??={ path: [] };
    this.rule_(520,$next);
    for(;;) switch(this.s) {
      case 520: this.fromPath({table:$.expr,category:'artref'},521); continue;
      case 521: switch(this.l()) {
        case ':': if(this.c(522)) { if (!$.expr.scope) $.expr.scope = $.expr.path.length; else {
this.warning( 'syntax-invalid-path-separator', this.lb(),
{ '#': 'colon', code: ':', newcode: '.' } );
} } continue;
        default: this.s=523; continue;
      }
      case 522: this.fromPath({table:$.expr,category:'ref'},523); continue;
      case 523: switch(this.lk()) {
        case 'as': this.ck(524); continue;
        case 'Id': if(this.gc(525,'tableAlias') && this.ci(0,'FromAlias')) { $.expr.name = this.fragileAlias(); } continue;
        default: this.giR(525); continue;
      }
      case 524: if(this.mi(0,'FromAlias')) { $.expr.name = this.identAst(); } continue;
      case 525: if(this.gr([])) { this.classifyImplicitName( $.expr.scope ? 'FromElemImplicit' : 'FromImplicit', $.expr ); } continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  fromPath($,$next) {
    this.rule_(527,$next);
    let pathItem;
    for(;;) switch(this.s) {
      case 527: if(this.mi(528,$.category)) { $.table.path.push( pathItem = this.identAst() ); } continue;
      case 528: switch(this.l()) {
        case '(': case '[': if(this.fromArgumentsAndFilter({pathStep:pathItem},529)) { pathItem = null; } continue;
        default: this.s=529; continue;
      }
      case 529: switch(this.l()) {
        case '.': if(this.c(530)) { if (!pathItem && !$.table.scope) {
$.table.scope = $.table.path.length; $.category = 'ref';
this.warning( 'syntax-invalid-path-separator', this.lb(),
{ '#': 'dot', code: '.', newcode: ':' } );
} } continue;
        default: this.gr(['Id','as']); continue;
      }
      case 530: if(this.miA(531,$.category)) { $.table.path.push( pathItem = this.identAst() ); } continue;
      case 531: switch(this.l()) {
        case '(': case '[': if(this.fromArgumentsAndFilter({pathStep:pathItem},529)) { pathItem = null; } continue;
        default: this.s=529; continue;
      }
      default:
      this.attachLocation( $.table.path );
      return this.exit_();
    }
  }
  fromArgumentsAndFilter($,$next) {
    this.rule_(533,$next);
    for(;;) switch(this.s) {
      case 533: switch(this.l()) {
        case '(': if(this.c(534)) { $.pathStep.args = this.createDict(); $.pathStep.$syntax = ':'; } continue;
        default: this.s=538; continue;
      }
      case 534: switch(this.lk()) {
        case 'Id': this.s=535; continue;
        default: this.ei(); continue;
      }
      case 535: switch(this.l()) {
        case 'Id': this.fromNamedArgument($,536); continue;
        default: this.s=537; continue;
      }
      case 536: switch(this.l()) {
        case ',': this.c(535); continue;
        case ')': this.s=537; continue;
        default: this.e(); continue;
      }
      case 537: this.m(538,')'); continue;
      case 538: switch(this.l()) {
        case '[': this.cardinalityAndFilter($,0); continue;
        default: this.gr(['Id','.','as']); continue;
      }
      default: return this.exit_();
    }
  }
  fromNamedArgument($,$next) {
    this.rule_(540,$next);
    let name; let expr; let _;
    for(;;) switch(this.s) {
      case 540: if(this.mi(541,'paramname')) name=this.lb(); continue;
      case 541: this.m(542,':'); continue;
      case 542: if(this.expression(_={},0)) {expr=_.expr; this.addDef( expr, $.pathStep, 'args', 0, this.identAst( name ) ); } continue;
      default: return this.exit_();
    }
  }
  cardinalityAndFilter($,$next) {
    this.rule_(544,$next);
    for(;;) switch(this.s) {
      case 544: this.m(545,'['); continue;
      case 545: switch(this.l()) {
        case 'Number': if(this.gc(547,'beforeColon') && this.c(546)) { $.pathStep.cardinality = { targetMax: this.unsignedIntegerLiteral(), location: this.lb().location }; } continue;
        default: this.s=547; continue;
      }
      case 546: this.m(547,':'); continue;
      case 547: this.filterClauses({pathStep:$.pathStep},548); continue;
      case 548: this.m(0,']'); continue;
      default: return this.exit_();
    }
  }
  filterClauses($,$next) {
    this.rule_(550,$next);
    let cond; let _;
    for(;;) switch(this.s) {
      case 550: switch(this.lk()) {
        case 'where': this.ck(551); continue;
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': this.s=551; continue;
        case 'group': case 'limit': case 'order': case 'having': this.gP(552); continue;
        default: this.gi(552); continue;
      }
      case 551: if(this.condition(_={},552)) {cond=_.expr; $.pathStep.where = cond; } continue;
      case 552: switch(this.lk()) {
        case 'group': this.s=553;{ this.csnParseOnly('syntax-unexpected-sql-clause', 1, { keyword: 'group by' }); } continue;
        default: this.s=554; continue;
      }
      case 553: this.groupByClause({query:$.pathStep},554); continue;
      case 554: switch(this.lk()) {
        case 'having': if(this.ck(555)) { this.csnParseOnly('syntax-unexpected-sql-clause', -1, { keyword: 'having' }); } continue;
        default: this.s=556; continue;
      }
      case 555: if(this.condition(_={},556)) {cond=_.expr; $.pathStep.having = cond; } continue;
      case 556: switch(this.lk()) {
        case 'limit': case 'order': this.s=557;{ if (this.lk() === 'limit') this.csnParseOnly('syntax-unexpected-sql-clause', 0, { keyword: 'limit' } ); else this.csnParseOnly('syntax-unexpected-sql-clause', 1, { keyword: 'order by' } ); } continue;
        default: this.gr([']']); continue;
      }
      case 557: this.orderByLimitOffset({query:$.pathStep},0); continue;
      default: return this.exit_();
    }
  }
  excludingClause($,$next) {
    this.rule_(559,$next);
    for(;;) switch(this.s) {
      case 559: this.mk(560,'excluding'); continue;
      case 560: if(this.m(561,'{')) { $.query.excludingDict = this.createDict(); } continue;
      case 561: switch(this.l()) {
        case 'Id': this.s=562; continue;
        default: this.e(); continue;
      }
      case 562: switch(this.l()) {
        case 'Id': if(this.ciA(563,'ref')) { this.addDef( { location: this.lb().location }, $.query, 'excludingDict', '', this.identAst() ); } continue;
        default: this.s=564; continue;
      }
      case 563: switch(this.l()) {
        case ',': this.c(562); continue;
        case '}': this.s=564; continue;
        default: this.e(); continue;
      }
      case 564: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.query.excludingDict ); } continue;
      default: return this.exit_();
    }
  }
  selectItemsList($,$next) {
    this.rule_(566,$next);
    for(;;) switch(this.s) {
      case 566: if(this.m(567,'{')) {this.notInExpandInline(); $.query.columns = this.createArray( $.start ); } continue;
      case 567: switch(this.l()) {
        case '*': if(this.c(568)) { $.query.columns.push( this.valueWithLocation() ); } continue;
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case '@': case '{': case 'Number': case 'String': case 'QuotedLiteral': this.selectItemDef({columns:$.query.columns},568); continue;
        default: this.s=569; continue;
      }
      case 568: switch(this.l()) {
        case ',': this.c(567); continue;
        case '}': this.s=569; continue;
        default: this.e(); continue;
      }
      case 569: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.query.columns ); } continue;
      default: return this.exit_();
    }
  }
  nestedSelectItemsList($,$next) {
    this.rule_(571,$next);
    for(;;) switch(this.s) {
      case 571: if(this.m(572,'{')) {this.inExpandInline(); $.query[$.clause] = this.createArray(); } continue;
      case 572: switch(this.l()) {
        case '*': if(this.c(573)) { $.query[$.clause].push( this.valueWithLocation() ); } continue;
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case '@': case '{': case 'Number': case 'String': case 'QuotedLiteral': this.selectItemDef({columns:$.query[$.clause]},573); continue;
        default: this.s=574; continue;
      }
      case 573: switch(this.l()) {
        case ',': this.c(572); continue;
        case '}': this.s=574; continue;
        default: this.e(); continue;
      }
      case 574: if(this.m(0,'}')) {this.afterBrace(); this.finalizeDictOrArray( $.query[$.clause] ); } continue;
      default: return this.exit_();
    }
  }
  selectItemDef($,$next) {
    this.rule_(576,$next);
    let art=new XsnArtifact(); let alias;
    let expr; let as; let target; let cond; let assoc; let card; let _;
    { $.columns.push( art );
this.docComment( art ); }
    for(;;) switch(this.s) {
      case 576: switch(this.l()) {
        case '@': this.annoAssignCol({art},576); continue;
        default: this.s=577; continue;
      }
      case 577: switch(this.lk()) {
        case 'virtual': if(this.lP(['Id','#','(','+','-',':','?','{','key','not','case','cast','null','true','false','Number','String','exists','QuotedLiteral']) && this.gc(578,'notInExpandInline') && this.ck(578)) { art.virtual = this.valueWithLocation( true ); } continue;
        default: this.s=578; continue;
      }
      case 578: switch(this.lk()) {
        case 'key': if(this.gc(579,'notInExpandInline') && this.ck(579)) { art.key = this.valueWithLocation( true ); } continue;
        default: this.s=579; continue;
      }
      case 579: switch(this.l()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'Number': case 'String': case 'QuotedLiteral': if(this.expression(_={},580)) {expr=_.expr; art.value = expr; } continue;
        case '{': this.nestedSelectItemsList({query:art,clause:'expand'},587); continue;
        default: this.e(); continue;
      }
      case 580: switch(this.lk()) {
        case 'as': if(this.ck(581)) as=this.lb(); continue;
        case 'Id': if(this.ci(582,'ItemAlias')) { art.name = this.fragileAlias( true ); } continue;
        default: if(this.gi(582)) { alias = this.classifyImplicitName( 'ItemImplicit', expr ); } continue;
      }
      case 581: if(this.mi(582,'ItemAlias')) { art.name = this.identAst(); } continue;
      case 582: switch(this.l()) {
        case '{': this.s=583;{ this.reportExpandInline( art, false ); } continue;
        case '.': if(this.c(585)) { this.reportExpandInline( art, as || true );
if (alias) alias.token.parsedAs = alias.parsedAs; } continue;
        default: this.s=590; continue;
      }
      case 583: this.nestedSelectItemsList({query:art,clause:'expand'},584); continue;
      case 584: switch(this.lk()) {
        case 'excluding': this.excludingClause({query:art},590); continue;
        default: this.s=590; continue;
      }
      case 585: switch(this.l()) {
        case '{': this.nestedSelectItemsList({query:art,clause:'inline'},586); continue;
        case '*': if(this.c(590)) { art.inline = [ this.valueWithLocation() ]; } continue;
        default: this.e(); continue;
      }
      case 586: switch(this.lk()) {
        case 'excluding': this.excludingClause({query:art},590); continue;
        default: this.s=590; continue;
      }
      case 587: switch(this.lk()) {
        case 'excluding': this.excludingClause({query:art},588); continue;
        default: this.s=588; continue;
      }
      case 588: this.mk(589,'as'); continue;
      case 589: if(this.mi(590,'ItemAlias')) { art.name = this.identAst(); } continue;
      case 590: this.s=591;{ this.docComment( art ); } continue;
      case 591: switch(this.l()) {
        case '@': this.annoAssignMid({art},591); continue;
        default: this.s=592; continue;
      }
      case 592: switch(this.l()) {
        case ':': this.c(593); continue;
        default: this.gr([',']); continue;
      }
      case 593: switch(this.lk()) {
        case 'type': this.lP() && this.typeTypeOf({art},607); continue;
        case 'localized': if(this.ckP(594,['Id'])) { art.localized = this.valueWithLocation( true ); } continue;
        case 'Id': this.s=594; continue;
        case 'redirected': this.ckP(595,['to']); continue;
        case 'association': if(this.ckP(599,['[','to'])) {assoc=this.lb(); this.associationInSelectItem( art ); } continue;
        case 'composition': if(this.ckP(601,['[','of'])) {assoc=this.lb(); this.associationInSelectItem( art ); } continue;
        default: this.ei(); continue;
      }
      case 594: this.typeRefOptArgs({art},607); continue;
      case 595: this.mk(596,'to'); continue;
      case 596: if(this.simplePath(_={category:'artref'},597)) {target=_.ref; art.target = target; } continue;
      case 597: switch(this.lk()) {
        case 'on': this.ck(598); continue;
        case '{': this.foreignKeysBlock({art},607); continue;
        default: this.s=607; continue;
      }
      case 598: if(this.condition(_={},607)) {cond=_.expr; art.on = cond; } continue;
      case 599: switch(this.l()) {
        case '[': this.cardinality({art},600); continue;
        default: this.s=600; continue;
      }
      case 600: this.mk(603,'to'); continue;
      case 601: switch(this.l()) {
        case '[': this.cardinality({art},602); continue;
        default: this.s=602; continue;
      }
      case 602: this.mk(603,'of'); continue;
      case 603: switch(this.lk()) {
        case 'one': case 'many': if(this.ckP(604,['Id'])) card=this.lb(); continue;
        default: this.s=604; continue;
      }
      case 604: if(this.simplePath(_={category:'artref'},605)) {target=_.ref; this.setAssocAndComposition( art, assoc, card, target ); } continue;
      case 605: this.mk(606,'on'); continue;
      case 606: if(this.condition(_={},607)) {expr=_.expr; art.on = expr; } continue;
      case 607: this.s=608;{ this.docComment( art ); } continue;
      case 608: switch(this.l()) {
        case '@': this.annoAssignStd({art},608); continue;
        default: this.gr([',']); continue;
      }
      default:
      this.attachLocation( art );
      return this.exit_();
    }
  }
  whereGroupByHaving($,$next) {
    this.rule_(610,$next);
    let cond; let _;
    for(;;) switch(this.s) {
      case 610: switch(this.lk()) {
        case 'where': this.ck(611); continue;
        default: this.s=612; continue;
      }
      case 611: if(this.condition(_={},612)) {cond=_.expr; $.query.where = cond; } continue;
      case 612: switch(this.lk()) {
        case 'group': this.groupByClause($,613); continue;
        default: this.s=613; continue;
      }
      case 613: switch(this.lk()) {
        case 'having': this.ck(614); continue;
        default: this.gr(['limit','order']); continue;
      }
      case 614: if(this.condition(_={},0)) {cond=_.expr; $.query.having = cond; } continue;
      default: return this.exit_();
    }
  }
  groupByClause($,$next) {
    this.rule_(616,$next);
    let expr; let _;
    for(;;) switch(this.s) {
      case 616: this.mk(617,'group'); continue;
      case 617: this.mk(618,'by'); continue;
      case 618: if(this.expression(_={},619)) {expr=_.expr; $.query.groupBy = [ expr ]; } continue;
      case 619: switch(this.l()) {
        case ',': this.c(620); continue;
        default: this.gr(['limit','order','having']); continue;
      }
      case 620: if(this.expression(_={},619)) {expr=_.expr; $.query.groupBy.push( expr ); } continue;
      default: return this.exit_();
    }
  }
  orderByLimitOffset($,$next) {
    this.rule_(622,$next);
    let expr; let _;
    for(;;) switch(this.s) {
      case 622: switch(this.lk()) {
        case 'order': this.orderByClause($,623); continue;
        default: this.s=623; continue;
      }
      case 623: switch(this.lk()) {
        case 'limit': this.ck(624); continue;
        default: this.gr([]); continue;
      }
      case 624: if(this.expression(_={},625)) {expr=_.expr; $.query.limit = { rows: expr }; } continue;
      case 625: switch(this.lk()) {
        case 'offset': this.ck(626); continue;
        default: this.gr([]); continue;
      }
      case 626: if(this.expression(_={},0)) {expr=_.expr; $.query.limit.offset = expr; } continue;
      default: return this.exit_();
    }
  }
  orderByClause($,$next) {
    this.rule_(628,$next);
    let expr; let _;
    for(;;) switch(this.s) {
      case 628: this.mk(629,'order'); continue;
      case 629: this.mk(630,'by'); continue;
      case 630: if(this.orderByExpression(_={},631)) {expr=_.expr; $.query.orderBy = [ expr ]; } continue;
      case 631: switch(this.l()) {
        case ',': this.c(632); continue;
        default: this.gr(['limit']); continue;
      }
      case 632: if(this.orderByExpression(_={},631)) {expr=_.expr; $.query.orderBy.push( expr ); } continue;
      default: return this.exit_();
    }
  }
  orderByExpression($,$next) {
    $.expr??=undefined;
    this.rule_(634,$next);
    for(;;) switch(this.s) {
      case 634: this.expression($,635); continue;
      case 635: switch(this.lk()) {
        case 'asc': case 'desc': if(this.ck(636)) { $.expr.sort = this.valueWithLocation(); } continue;
        default: this.s=636; continue;
      }
      case 636: switch(this.lk()) {
        case 'nulls': this.ck(637); continue;
        default: this.gr([',','limit']); continue;
      }
      case 637: switch(this.lk()) {
        case 'last': case 'first': if(this.ck(0)) { $.expr.nulls = this.valueWithLocation(); } continue;
        default: this.e(); continue;
      }
      default: return this.exit_();
    }
  }
  conditionEOF($,$next) {
    $.cond??=undefined;
    this.rule_(639,$next);
    let _;
    for(;;) switch(this.s) {
      case 639: if(this.expression(_={},640)) $.cond=_.expr; continue;
      case 640: this.m(0,'EOF'); continue;
      default: return this.exit_();
    }
  }
  condition($,$next) {
    $.expr??=undefined;
    this.rule_(642,$next);
    for(;;) switch(this.s) {
      case 642: this.expression($,0); continue;
      default: return this.exit_();
    }
  }
  valuePath($,$next,$startState) {
    $.expr??={ path: [] };
    this.rule_($startState??644,$next);
    let pathItem;
    for(;;) switch(this.s) {
      case 644: if(this.mi(645,'ref')) { $.expr.path.push( pathItem = this.identAst() ); } continue;
      case 645: switch(this.l()) {
        case '(': case '[': this.argumentsAndFilter({pathStep:pathItem},646); continue;
        default: this.s=646; continue;
      }
      case 646: switch(this.l()) {
        case '.': this.gc(0,'isDotForPath') && this.c(647); continue;
        default: this.gr(['*','+','-','/','<','=','>','?','!=','<=','<>','>=','in','is','or','||','and','not','like','between']); continue;
      }
      case 647: if(this.miA(648,'ref')) { $.expr.path.push( pathItem = this.identAst() ); } continue;
      case 648: switch(this.l()) {
        case '(': case '[': if(this.argumentsAndFilter({pathStep:pathItem},646)) { pathItem = null; } continue;
        default: this.s=646; continue;
      }
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  expression($,$next,$startState) {
    $.expr??=undefined;
    this.rule_($startState??650,$next);
    let e; let open; let _;
    for(;;) switch(this.s) {
      case 650: switch(this.lk()) {
        case '(': this.expressionOrQueryParens($,651); continue;
        case '#': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'QuotedLiteral': this.literalValue($,661); continue;
        case ':': if(this.c(652)) { this.reportUnexpectedSpace(); } continue;
        case '?': if(this.c(661)) { this.csnParseOnly( 'syntax-unsupported-param', -1, { '#': 'dynamic', code: '?' } );
$.expr = this.attachLocation({ param: this.valueWithLocation(), scope: 'param' }); } continue;
        case 'Id': if(this.valuePath(_={},654)) {e=_.expr; $.expr = this.valuePathAst( e ); } continue;
        case 'new': this.lP() && this.newAndValuePath($,661); continue;
        case 'exists': if(this.ck(657)) { $.expr = this.applyOpToken(); } continue;
        case 'case': this.caseExpression($,661); continue;
        case 'cast': this.castFunction($,661); continue;
        case '+': case '-': if(this.ep(30,'right') && this.c(660)) { $.expr = this.applyOpToken(); } continue;
        case 'not': if(this.ep(8,'right') && this.ck(660)) { $.expr = this.applyOpToken(); } continue;
        default: this.ei(); continue;
      }
      case 651: this.s=661;{ $.expr = this.taggedIfQuery( $.expr ); } continue;
      case 652: switch(this.l()) {
        case 'Id': if(this.ciA(653,'paramref')) { $.expr = { path: [ this.identAst() ], location: this.startLocation(), scope: 'param' }; } continue;
        case 'Number': if(this.c(661)) { this.csnParseOnly( 'syntax-unsupported-param', -1, { '#': 'positional', code: ':' + this.lb().text } );
$.expr = this.attachLocation({ param: this.unsignedIntegerLiteral(), scope: 'param' }); } continue;
        default: this.e(); continue;
      }
      case 653: switch(this.l()) {
        case '.': if(this.valuePath($,661,646)) { $.expr = this.valuePathAst( $.expr ); } continue;
        default: this.s=661;{ this.attachLocation( $.expr ); } continue;
      }
      case 654: switch(this.lk()) {
        case 'over': if(this.ckP(655,['('])) { this.pushXprToken( $.expr.suffix = [] ); } continue;
        default: this.s=656; continue;
      }
      case 655: this.overClause(e={outer:$.expr.suffix},656); continue;
      case 656: this.s=661;{ this.attachLocation( $.expr ); } continue;
      case 657: switch(this.l()) {
        case '(': if(this.c(658)) open=this.lb(); continue;
        case 'Id': if(this.valuePath(_={},661)) {e=_.expr; e = this.valuePathAst( e ); e.$expected = 'exists';
$.expr.args.push( e ); this.attachLocation( $.expr ); } continue;
        case '?': if(this.c(661)) { this.csnParseOnly( 'syntax-unsupported-param', -1, { '#': 'dynamic', code: '?' } );
$.expr.args.push( { param: this.valueWithLocation(), scope: 'param' } ); this.attachLocation( $.expr ); } continue;
        default: this.e(); continue;
      }
      case 658: if(this.queryExpression(_={},659)) e=_.expr; continue;
      case 659: if(this.m(661,')')) { $.expr.args.push( this.taggedIfQuery( this.surroundByParens( e, open ) ) );
this.attachLocation( $.expr ); } continue;
      case 660: if(this.expression(_={},661)) {e=_.expr; $.expr = this.signedExpression( $.expr, e ); } continue;
      case 661: switch(this.lk()) {
        case '*': case '/': if(this.gp(0,24) && this.c(665)) { $.expr = this.applyOpToken( $.expr, 'nary' ); } continue;
        case '+': case '-': if(this.gp(0,22) && this.c(665)) { $.expr = this.applyOpToken( $.expr, 'nary' ); } continue;
        case '||': if(this.gp(0,20) && this.c(665)) { $.expr = this.applyOpToken( $.expr, 'nary' ); } continue;
        case 'and': if(this.lP(['Id','#','(','+','-',':','?','not','case','cast','null','true','false','Number','String','exists','QuotedLiteral']) && this.gp(0,4) && this.ck(665)) { $.expr = this.applyOpToken( $.expr, 'nary' ); } continue;
        case 'or': if(this.lP(['Id','#','(','+','-',':','?','not','case','cast','null','true','false','Number','String','exists','QuotedLiteral']) && this.gp(0,2) && this.ck(665)) { $.expr = this.applyOpToken( $.expr, 'nary' ); } continue;
        case '?': if(this.gp(0,0) && this.c(662)) { $.expr = this.applyOpToken( $.expr, '?:' ); } continue;
        case '<': case '=': case '>': case '!=': case '<=': case '<>': case '>=': if(this.gp(0,10,'none') && this.c(664)) { $.expr = this.applyOpToken( $.expr ); } continue;
        case 'is': if(this.lP(['not','null']) && this.gp(0,10,'none') && this.ck(666)) { $.expr = this.applyOpToken( $.expr ); } continue;
        case 'not': if(this.gc(0,'isNegatedRelation') && this.ck(668)) { $.expr = this.applyOpToken( $.expr ); } continue;
        case 'in': case 'like': case 'between': if(this.lP() && this.gp(0,10,'none') && this.g(668)) { $.expr = { op: { val: 'ixpr', location: this.la().location }, args: [ $.expr ] }; } continue;
        default: this.gr([]); continue;
      }
      case 662: if(this.expression(_={},663)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 663: if(this.m(665,':')) { this.pushXprToken( $.expr ); } continue;
      case 664: switch(this.lk()) {
        case 'all': case 'any': case 'some': if(this.ck(665)) { this.pushXprToken( $.expr ); } continue;
        default: this.s=665; continue;
      }
      case 665: if(this.expression(_={},676)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 666: switch(this.lk()) {
        case 'not': if(this.ck(667)) { this.pushXprToken( $.expr ); } continue;
        default: this.s=667; continue;
      }
      case 667: if(this.mk(676,'null')) { this.pushXprToken( $.expr ); } continue;
      case 668: switch(this.lk()) {
        case 'between': if(this.ck(669)) { this.pushXprToken( $.expr ); } continue;
        case 'in': if(this.ck(672)) { this.pushXprToken( $.expr ); } continue;
        case 'like': if(this.ck(673)) { this.pushXprToken( $.expr ); } continue;
        default: this.e(); continue;
      }
      case 669: if(this.expression(_={},670)) {e=_.expr; $.expr.args?.push( e ); } continue;
      case 670: if(this.mk(671,'and')) { this.pushXprToken( $.expr ); } continue;
      case 671: if(this.expression(_={},676)) {e=_.expr; $.expr.args?.push( e ); } continue;
      case 672: if(this.expression(_={},676)) {e=_.expr; $.expr.args?.push( this.secureParens( e ) ); } continue;
      case 673: if(this.expression(_={},674)) {e=_.expr; $.expr.args?.push( e ); } continue;
      case 674: switch(this.lk()) {
        case 'escape': if(this.ckP(675,['Id','#','(','+','-',':','?','not','case','cast','null','true','false','Number','String','exists','QuotedLiteral'])) { this.pushXprToken( $.expr ); } continue;
        default: this.s=676; continue;
      }
      case 675: if(this.expression(_={},676)) {e=_.expr; $.expr.args?.push( e ); } continue;
      case 676: this.s=661;{ this.attachLocation( $.expr ); } continue;
      default: return this.exit_();
    }
  }
  expressionOrQueryParens($,$next) {
    $.expr??=undefined;
    this.rule_(678,$next);
    for(;;) switch(this.s) {
      case 678: this.m(679,'('); continue;
      case 679: switch(this.lk()) {
        case '(': this.expressionOrQueryParens($,680); continue;
        case 'Id': case '#': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': this.ep(-2) && this.expression($,682); continue;
        case 'select': this.ep(-1) && this.queryExpression($,683); continue;
        default: this.ei(); continue;
      }
      case 680: switch(this.lk()) {
        case '*': case '+': case '-': case '/': case '<': case '=': case '>': case '?': case '!=': case '<=': case '<>': case '>=': case 'in': case 'is': case 'or': case '||': case 'and': case 'not': case 'like': case 'between': this.gp(683,-2,'post') && this.expression($,681,651); continue;
        case ',': this.gp(683,-2,'post') && this.continueExpressionslist($,683); continue;
        case 'limit': case 'minus': case 'order': case 'union': case 'except': case 'intersect': this.gp(683,-1,'post') && this.queryExpression($,683,464); continue;
        default: this.s=683; continue;
      }
      case 681: switch(this.l()) {
        case ',': this.continueExpressionslist($,683); continue;
        default: this.s=683; continue;
      }
      case 682: switch(this.l()) {
        case ',': this.continueExpressionslist($,683); continue;
        default: this.s=683; continue;
      }
      case 683: if(this.m(0,')')) {this.setPrecInCallingRule(); this.surroundByParens( $.expr ); } continue;
      default: return this.exit_();
    }
  }
  continueExpressionslist($,$next) {
    this.rule_(685,$next);
    let e; let _;
    for(;;) switch(this.s) {
      case 685: if(this.m(686,',')) { $.expr = { op: this.valueWithLocation( 'list' ), args: [ $.expr ], location: { ... $.expr.$parens?.at( -1 ) ?? $.expr.location } }; } continue;
      case 686: switch(this.lk()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': this.s=687; continue;
        default: this.ei(); continue;
      }
      case 687: switch(this.lk()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': if(this.expression(_={},688)) {e=_.expr; $.expr.args.push( e ); } continue;
        default: this.gi(0,[')']); continue;
      }
      case 688: switch(this.l()) {
        case ',': this.c(687); continue;
        default: this.gr([')']); continue;
      }
      case 0: { this.attachLocation( $.expr ); }
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  newAndValuePath($,$next) {
    $.expr??=undefined;
    this.rule_(690,$next);
    let e; let _;
    for(;;) switch(this.s) {
      case 690: if(this.mk(691,'new')) { $.expr = this.applyOpToken(); } continue;
      case 691: if(this.valuePath(_={},0)) {e=_.expr; e = this.valuePathAst( e );
if (e.op?.val !== 'ixpr') $.expr.args.push( e ); else $.expr.args.push( ...e.args ); } continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  caseExpression($,$next) {
    $.expr??={ op: { val: 'ixpr' }, args: [] };
    this.rule_(693,$next);
    let e; let _;
    for(;;) switch(this.s) {
      case 693: if(this.mk(694,'case')) { this.pushXprToken( $.expr ); $.expr.op.location = $.expr.args[0].location; } continue;
      case 694: switch(this.lk()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': if(this.expression(_={},695)) {e=_.expr; $.expr.args.push( e ); } continue;
        case 'when': this.gP(695); continue;
        default: this.gi(695); continue;
      }
      case 695: switch(this.lk()) {
        case 'when': this.s=696; continue;
        default: this.e(); continue;
      }
      case 696: switch(this.lk()) {
        case 'when': if(this.ck(697)) { this.pushXprToken( $.expr ); } continue;
        default: this.s=700; continue;
      }
      case 697: if(this.expression(_={},698)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 698: if(this.mk(699,'then')) { this.pushXprToken( $.expr ); } continue;
      case 699: if(this.expression(_={},696)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 700: switch(this.lk()) {
        case 'else': if(this.ck(701)) { this.pushXprToken( $.expr ); } continue;
        default: this.s=702; continue;
      }
      case 701: if(this.expression(_={},702)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 702: if(this.mk(0,'end')) { this.pushXprToken( $.expr ); } continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  castFunction($,$next) {
    $.expr??={};
    this.rule_(704,$next);
    let arg; let _;
    for(;;) switch(this.s) {
      case 704: if(this.mk(705,'cast')) { $.expr.op = this.valueWithLocation(); } continue;
      case 705: if(this.m(706,'(')) { $.expr.args = this.createArray(); } continue;
      case 706: if(this.expression(_={},707)) {arg=_.expr; $.expr.args?.push( arg ); } continue;
      case 707: this.mk(708,'as'); continue;
      case 708: this.typeRefOptArgs({art:$.expr},709); continue;
      case 709: if(this.m(0,')')) { this.finalizeDictOrArray( $.expr.args ); } continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  argumentsAndFilter($,$next) {
    this.rule_(711,$next);
    let open; let expr; let id; let _;
    for(;;) switch(this.s) {
      case 711: switch(this.l()) {
        case '(': if(this.c(712)) {open=this.lb();this.prepareSpecialFunction(); $.pathStep.args = this.createArray(); } continue;
        default: this.s=731; continue;
      }
      case 712: switch(this.l()) {
        case 'Id': if(this.gc(713,'isNamedArg') && this.ciA(719,'paramname')) id=this.lb(); continue;
        default: this.s=713; continue;
      }
      case 713: switch(this.l()) {
        case 'Id': case '#': case '(': case '*': case '+': case '-': case ':': case '?': case 'Id_all': case 'Number': case 'String': case 'QuotedLiteral': if(this.funcExpression(_={},714)) {expr=_.expr; $.pathStep.args.push( expr ); } continue;
        default: this.s=730; continue;
      }
      case 714: switch(this.l()) {
        case ',': if(this.c(715)) {this.nextFunctionArgument();} continue;
        default: this.s=716; continue;
      }
      case 715: switch(this.lk()) {
        case 'Id': case '#': case '(': case '*': case '+': case '-': case ':': case '?': case 'not': case 'case': case 'cast': case 'null': case 'true': case 'false': case 'Id_all': case 'Number': case 'String': case 'exists': case 'QuotedLiteral': if(this.funcExpression(_={},714)) {expr=_.expr; $.pathStep.args.push( expr ); } continue;
        case ')': case 'order': this.gP(716); continue;
        default: this.ei(); continue;
      }
      case 716: switch(this.lk()) {
        case 'order': if(this.ck(717)) { expr = $.pathStep.args[$.pathStep.args.length - 1] = this.applyOpToken( expr ); } continue;
        default: this.s=730; continue;
      }
      case 717: if(this.mk(718,'by')) { this.pushXprToken( expr ); } continue;
      case 718: if(this.orderByClauseAsXpr({outer:expr.args},730)) { this.attachLocation( expr ); } continue;
      case 719: switch(this.l()) {
        case ':': if(this.c(720)) { $.pathStep.args = this.createDict( open ); $.pathStep.$syntax = ':'; } continue;
        case '=>': if(this.c(725)) { $.pathStep.args = this.createDict(); } continue;
        default: this.e(); continue;
      }
      case 720: if(this.expression(_={},721)) {expr=_.expr; this.addNamedArg( $.pathStep, id, expr ); } continue;
      case 721: switch(this.l()) {
        case ',': this.c(722); continue;
        default: this.s=730; continue;
      }
      case 722: switch(this.l()) {
        case 'Id': if(this.ciA(723,'paramname')) id=this.lb(); continue;
        case ')': this.s=730; continue;
        default: this.e(); continue;
      }
      case 723: this.m(724,':'); continue;
      case 724: if(this.expression(_={},721)) {expr=_.expr; this.addNamedArg( $.pathStep, id, expr ); } continue;
      case 725: if(this.expression(_={},726)) {expr=_.expr; this.addNamedArg( $.pathStep, id, expr ); } continue;
      case 726: switch(this.l()) {
        case ',': this.c(727); continue;
        default: this.s=730; continue;
      }
      case 727: switch(this.l()) {
        case 'Id': if(this.ciA(728,'paramname')) id=this.lb(); continue;
        case ')': this.s=730; continue;
        default: this.e(); continue;
      }
      case 728: this.m(729,'=>'); continue;
      case 729: if(this.expression(_={},726)) {expr=_.expr; this.addNamedArg( $.pathStep, id, expr ); } continue;
      case 730: this.m(731,')'); continue;
      case 731: switch(this.l()) {
        case '[': this.cardinalityAndFilter($,0); continue;
        default: this.gr(['*','+','-','.','/','<','=','>','?','!=','<=','<>','>=','in','is','or','||','and','not','like','between']); continue;
      }
      default: return this.exit_();
    }
  }
  funcExpression($,$next) {
    $.expr??=undefined;
    this.rule_(733,$next);
    let args;
    let tok; let e; let _;
    for(;;) switch(this.s) {
      case 733: switch(this.lGenericIntroOrExpr()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'Number': case 'String': case 'QuotedLiteral': if(this.expression(_={},735)) $.expr=_.expr; continue;
        case 'GenericExpr': if(this.ckA(735)) {tok=this.lb(); $.expr = { val: tok.keyword ?? tok.type, location: tok.location, literal: 'token' }; } continue;
        case 'GenericIntro': if(this.ckA(734)) { $.expr = this.applyOpToken(); args = $.expr.args; } continue;
        default: this.e(); continue;
      }
      case 734: if(this.expression(_={},735)) {e=_.expr; $.expr.args.push( e ); } continue;
      case 735: switch(this.lGenericSeparator()) {
        case 'GenericSeparator': if(this.ckA(736)) { if (args) this.pushXprToken( args ); else { $.expr= this.applyOpToken( $.expr ); args = $.expr.args; } } continue;
        default: this.gr([')',',','order']); continue;
      }
      case 736: switch(this.lGenericExpr()) {
        case 'Id': case '#': case '(': case '+': case '-': case ':': case '?': case 'Number': case 'String': case 'QuotedLiteral': if(this.expression(_={},735)) {e=_.expr; args.push( e ); } continue;
        case 'GenericExpr': if(this.ckA(735)) { this.pushXprToken( args ); } continue;
        default: this.e(); continue;
      }
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  overClause($,$next) {
    this.rule_(738,$next);
    let over=[];
    for(;;) switch(this.s) {
      case 738: this.m(739,'('); continue;
      case 739: switch(this.lk()) {
        case 'partition': if(this.ck(740)) { this.pushXprToken( over ); } continue;
        default: this.s=742; continue;
      }
      case 740: if(this.mk(741,'by')) { this.pushXprToken( over ); } continue;
      case 741: this.expressionsAsXpr({outer:over},742); continue;
      case 742: switch(this.lk()) {
        case 'order': if(this.ck(743)) { this.pushXprToken( over ); } continue;
        default: this.s=745; continue;
      }
      case 743: if(this.mk(744,'by')) { this.pushXprToken( over ); } continue;
      case 744: this.orderByClauseAsXpr({outer:over},745); continue;
      case 745: switch(this.lk()) {
        case 'rows': if(this.ck(746)) { this.pushXprToken( over ); } continue;
        default: this.s=747; continue;
      }
      case 746: this.windowFrameClause({outer:over},747); continue;
      case 747: this.m(0,')'); continue;
      default:
      $.outer.push( this.surroundByParens( this.ixprAst( over ) ) );
      return this.exit_();
    }
  }
  expressionsAsXpr($,$next) {
    this.rule_(749,$next);
    let args=[];
    let expr; let _;
    for(;;) switch(this.s) {
      case 749: if(this.expression(_={},750)) {expr=_.expr; args.push( expr ); } continue;
      case 750: switch(this.l()) {
        case ',': if(this.c(751)) { this.pushXprToken( args ); } continue;
        default: this.gr([')','rows','order']); continue;
      }
      case 751: if(this.expression(_={},750)) {expr=_.expr; args.push( expr ); } continue;
      default:
      $.outer.push( this.ixprAst( args ) );
      return this.exit_();
    }
  }
  orderByClauseAsXpr($,$next) {
    this.rule_(753,$next);
    let args=[];
    for(;;) switch(this.s) {
      case 753: this.orderBySpecAsXpr({args},754); continue;
      case 754: switch(this.l()) {
        case ',': if(this.c(755)) { this.pushXprToken( args ); } continue;
        default: this.gr([')']); continue;
      }
      case 755: this.orderBySpecAsXpr({args},754); continue;
      default:
      $.outer.push( this.ixprAst( args ) );
      return this.exit_();
    }
  }
  orderBySpecAsXpr($,$next) {
    this.rule_(757,$next);
    let expr; let _;
    for(;;) switch(this.s) {
      case 757: if(this.expression(_={},758)) {expr=_.expr; $.args.push( expr ); } continue;
      case 758: switch(this.lk()) {
        case 'asc': case 'desc': if(this.ck(759)) { this.pushXprToken( $.args ); } continue;
        default: this.s=759; continue;
      }
      case 759: switch(this.lk()) {
        case 'nulls': if(this.ck(760)) { this.pushXprToken( $.args ); } continue;
        default: this.gr([')',',']); continue;
      }
      case 760: switch(this.lk()) {
        case 'last': case 'first': if(this.ck(0)) { this.pushXprToken( $.args ); } continue;
        default: this.e(); continue;
      }
      default: return this.exit_();
    }
  }
  windowFrameClause($,$next) {
    this.rule_(762,$next);
    let args=[];
    for(;;) switch(this.s) {
      case 762: switch(this.lk()) {
        case 'unbounded': if(this.ck(763)) { this.pushXprToken( args ); } continue;
        case 'Number': if(this.c(763)) { args.push( this.unsignedIntegerLiteral() ); } continue;
        case 'current': if(this.ck(764)) { this.pushXprToken( args ); } continue;
        case 'between': if(this.ck(765)) { this.pushXprToken( args ); } continue;
        default: this.e(); continue;
      }
      case 763: if(this.mk(0,'preceding')) { this.pushXprToken( args ); } continue;
      case 764: if(this.mk(0,'row')) { this.pushXprToken( args ); } continue;
      case 765: this.windowFrameBoundSpec({args},766); continue;
      case 766: if(this.mk(767,'and')) { this.pushXprToken( args ); } continue;
      case 767: this.windowFrameBoundSpec({args},0); continue;
      default:
      $.outer.push( this.ixprAst( args ) );
      return this.exit_();
    }
  }
  windowFrameBoundSpec($,$next) {
    this.rule_(769,$next);
    for(;;) switch(this.s) {
      case 769: switch(this.lk()) {
        case 'unbounded': if(this.ck(770)) { this.pushXprToken( $.args ); } continue;
        case 'Number': if(this.c(770)) { $.args.push( this.unsignedIntegerLiteral() ); } continue;
        case 'current': if(this.ck(772)) { this.pushXprToken( $.args ); } continue;
        default: this.e(); continue;
      }
      case 770: switch(this.lk()) {
        case 'following': this.ck(771); continue;
        case 'preceding': this.ck(771); continue;
        default: this.e(); continue;
      }
      case 771: if(this.gr([])) { this.pushXprToken( $.args ); } continue;
      case 772: if(this.mk(0,'row')) { this.pushXprToken( $.args ); } continue;
      default: return this.exit_();
    }
  }
  literalValue($,$next) {
    $.expr??={};
    this.rule_(774,$next);
    for(;;) switch(this.s) {
      case 774: switch(this.lk()) {
        case '#': if(this.c(775)) { this.reportUnexpectedSpace(); } continue;
        case 'null': if(this.ck(0)) { $.expr = { literal: 'null', val: null }; } continue;
        case 'true': case 'false': if(this.ck(0)) { $.expr = { literal: 'boolean', val: this.lb().keyword === 'true' }; } continue;
        case 'Number': if(this.c(0)) { $.expr = this.numberLiteral(); } continue;
        case 'String': if(this.c(0)) { $.expr = this.quotedLiteral(); } continue;
        case 'QuotedLiteral': if(this.c(0)) { $.expr = this.quotedLiteral(); } continue;
        default: this.e(); continue;
      }
      case 775: if(this.mi(0,'enumref')) { $.expr = { literal: 'enum', sym: this.identAst() } } continue;
      default:
      this.attachLocation( $.expr );
      return this.exit_();
    }
  }
  annoAssignStd($,$next) {
    this.rule_(777,$next);
    for(;;) switch(this.s) {
      case 777: if(this.m(778,'@')) {this.annoInSameLine(); this.reportUnexpectedSpace(); } continue;
      case 778: switch(this.l()) {
        case '(': this.annoAssignParen($,0); continue;
        case 'Id': this.annoAssignBase($,0); continue;
        default: this.e(); continue;
      }
      default:
      this.docComment( $.art );
      return this.exit_();
    }
  }
  annoAssignCol($,$next) {
    this.rule_(780,$next);
    for(;;) switch(this.s) {
      case 780: if(this.m(781,'@')) { this.reportUnexpectedSpace(); } continue;
      case 781: switch(this.l()) {
        case '(': this.annoAssignParen($,0); continue;
        case 'Id': this.annoAssignBase($,0); continue;
        default: this.e(); continue;
      }
      default:
      this.docComment( $.art );
      return this.exit_();
    }
  }
  annoAssignMid($,$next) {
    this.rule_(783,$next);
    let name; let _;
    for(;;) switch(this.s) {
      case 783: if(this.m(784,'@')) {this.annoInSameLine(); this.reportUnexpectedSpace(); } continue;
      case 784: switch(this.l()) {
        case '(': this.annoAssignParen($,0); continue;
        case 'Id': if(this.annoNamePath(_={category:'anno'},0)) {name=_.name; this.assignAnnotation( $.art, {}, name ); this.warnIfColonFollows( name ); } continue;
        default: this.e(); continue;
      }
      default:
      this.docComment( $.art );
      return this.exit_();
    }
  }
  annoAssignParen($,$next) {
    this.rule_(786,$next);
    for(;;) switch(this.s) {
      case 786: if(this.m(787,'(')) {this.annoInSameLine();} continue;
      case 787: switch(this.l()) {
        case 'Id': this.annoAssignBase({art:$.art},788); continue;
        default: this.s=789; continue;
      }
      case 788: switch(this.l()) {
        case ',': this.c(787); continue;
        case ')': this.s=789; continue;
        default: this.e(); continue;
      }
      case 789: this.m(0,')'); continue;
      default: return this.exit_();
    }
  }
  annoAssignBase($,$next) {
    this.rule_(791,$next);
    let value={};
    let name; let _;
    for(;;) switch(this.s) {
      case 791: if(this.annoNamePath(_={category:'anno'},792)) name=_.name; continue;
      case 792: switch(this.l()) {
        case ':': this.gc(0,'annoInSameLine') && this.c(793); continue;
        default: this.gr([]); continue;
      }
      case 793: if(this.annoValue(_={},0)) value=_.value; continue;
      default:
      this.assignAnnotation( $.art, value, name || {} );
      return this.exit_();
    }
  }
  annoNamePath($,$next) {
    $.name??=new XsnName();
    this.rule_(795,$next);
    let at;
    for(;;) switch(this.s) {
      case 795: if(this.miA(796,$.category)) { $.name.path = [ this.identAst() ]; } continue;
      case 796: switch(this.l()) {
        case '.': this.c(797); continue;
        default: this.s=799; continue;
      }
      case 797: switch(this.l()) {
        case 'Id': if(this.ciA(796,$.category)) { $.name.path.push( this.identAst() ); } continue;
        case '@': if(this.c(798)) at=this.lb(); continue;
        default: this.e(); continue;
      }
      case 798: if(this.miA(796,$.category)) { $.name.path.push( this.identAstWithPrefix( at ) ); } continue;
      case 799: switch(this.l()) {
        case '#': this.gc(0,'annoInSameLine') && this.annoPathVariant({nameOrRef:$.name},0); continue;
        default: this.gr([]); continue;
      }
      default:
      this.attachLocation( $.name );
      return this.exit_();
    }
  }
  annoPath($,$next) {
    this.rule_(801,$next);
    let at;
    for(;;) switch(this.s) {
      case 801: switch(this.l()) {
        case 'Id': if(this.ciA(803,$.category)) { $.nameOrRef.path = [ this.identAst() ]; } continue;
        case '@': if(this.c(802)) at=this.lb(); continue;
        default: this.e(); continue;
      }
      case 802: if(this.miA(803,$.category)) { $.nameOrRef.path = [ this.identAstWithPrefix( at ) ]; } continue;
      case 803: switch(this.l()) {
        case '.': this.c(804); continue;
        default: this.s=806; continue;
      }
      case 804: switch(this.l()) {
        case 'Id': if(this.ciA(803,$.category)) { $.nameOrRef.path.push( this.identAst() ); } continue;
        case '@': if(this.c(805)) at=this.lb(); continue;
        default: this.e(); continue;
      }
      case 805: if(this.miA(803,$.category)) { $.nameOrRef.path.push( this.identAstWithPrefix( at ) ); } continue;
      case 806: switch(this.l()) {
        case '#': this.annoPathVariant({nameOrRef:$.nameOrRef},0); continue;
        default: this.gr([]); continue;
      }
      default:
      this.attachLocation( $.nameOrRef );
      return this.exit_();
    }
  }
  annoPathVariant($,$next) {
    this.rule_(808,$next);
    for(;;) switch(this.s) {
      case 808: if(this.m(809,'#')) { this.reportUnexpectedSpace(); } continue;
      case 809: if(this.miA(810,'variant')) { $.nameOrRef.variant = { path: [ this.identAst() ] }; } continue;
      case 810: switch(this.l()) {
        case '.': this.c(811); continue;
        default: this.gr([]); continue;
      }
      case 811: if(this.miA(810,'variant')) { $.nameOrRef.variant.path.push( this.identAst() ); } continue;
      default:
      this.attachLocation( $.nameOrRef.variant );
      return this.exit_();
    }
  }
  annoStructValue($,$next) {
    $.value??={};
    this.rule_(813,$next);
    let name=new XsnName();
    let _;
    for(;;) switch(this.s) {
      case 813: if(this.annoPath({nameOrRef:name,category:'name'},814)) { this.attachLocation( name ); } continue;
      case 814: switch(this.l()) {
        case ':': this.c(815); continue;
        default: if(this.gr([',','}'])) { this.attachLocation( $.value ); } continue;
      }
      case 815: if(this.annoValue(_={},0)) $.value=_.value; continue;
      default:
      $.value.name = name;
      return this.exit_();
    }
  }
  annoValue($,$next) {
    $.value??={};
    this.rule_(817,$next);
    let sign; let sub; let ellipsis; let upTo; let _;
    for(;;) switch(this.s) {
      case 817: switch(this.lk()) {
        case '#': case 'null': case 'true': case 'false': case 'Number': case 'String': case 'QuotedLiteral': if(this.literalValue(_={},0)) {$.value=_.expr; this.adjustAnnoNumber( $.value ); } continue;
        case '+': case '-': if(this.c(818)) sign=this.lb(); continue;
        case 'Id': case '@': this.annoPath({nameOrRef:$.value,category:'annoref'},0); continue;
        case '{': if(this.c(819)) { if (!this.dynamic_.arrayAnno) $.value.$flatten = [];
else { $.value.struct = Object.create(null); $.value.literal = 'struct'; } } continue;
        case '[': if(this.c(822)) {this.ellipsisRestriction(); $.value.val = []; $.value.literal = 'array' } continue;
        case '(': this.c(829); continue;
        default: this.ei(); continue;
      }
      case 818: if(this.m(0,'Number')) { this.adjustAnnoNumber( $.value = this.numberLiteral( sign ) ); } continue;
      case 819: switch(this.l()) {
        case 'Id': case '@': if(this.annoStructValue(_={},820)) {sub=_.value; if ($.value.$flatten) $.value.$flatten.push( sub );
else this.addDef( sub, $.value, 'struct', null, sub.name ); } continue;
        default: this.s=821; continue;
      }
      case 820: switch(this.l()) {
        case ',': this.c(819); continue;
        case '}': this.s=821; continue;
        default: this.e(); continue;
      }
      case 821: this.m(0,'}'); continue;
      case 822: switch(this.l()) {
        case 'Id': case '#': case '(': case '+': case '-': case '@': case '[': case '{': case 'Number': case 'String': case 'QuotedLiteral': if(this.annoValue(_={},827)) {sub=_.value; $.value.val.push( sub ) } continue;
        case '...': if(this.gc(828,'ellipsisRestriction') && this.c(823)) ellipsis=this.lb(); continue;
        default: this.s=828; continue;
      }
      case 823: switch(this.lk()) {
        case 'up': this.ck(824); continue;
        default: this.s=826;{ upTo = undefined; } continue;
      }
      case 824: this.mk(825,'to'); continue;
      case 825: if(this.annoValue(_={},826)) upTo=_.value; continue;
      case 826: this.s=827;{ $.value.val.push( { literal: 'token', val: '...', location: ellipsis.location, upTo: upTo } ); } continue;
      case 827: switch(this.l()) {
        case ',': this.c(822); continue;
        case ']': this.s=828; continue;
        default: this.e(); continue;
      }
      case 828: this.m(0,']'); continue;
      case 829: if(this.condition(_={},830)) $.value=_.expr; continue;
      case 830: if(this.m(0,')')) { $.value.$tokenTexts = this.ruleTokensText(); } continue;
      default:
      this.attachLocation( $.value );
      return this.exit_();
    }
  }
}
module.exports = CdlParser;
